<!doctype html>
<html>
   <head>
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
      <meta name="generator" content="r-universe/pandoc">
      
      
      <title>Basic Usage &amp; Preprocessing</title>
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.6.0/css/bootstrap.min.css" data-external="1">
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" data-external="1">
      <link rel="stylesheet" href="https://r-universe.dev/static/article-styles.css?nocache=1" data-external="1">
   </head>
   <body>
      <div class="container" id="topcontainer">
         <script>if (window.self != window.top) try {document.getElementById('topcontainer').className = "container-fluid article-in-iframe"; } catch(e) {}</script>

         <div class="frontmatter">
         <h1 class="title">Basic Usage &amp; Preprocessing</h1>
         </div>

                  <div class="row">
            <div class="col-xl-10"><div id="preparation" class="section level1">
<h1>Preparation</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(sesame)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="do">## As sesame and sesameData are under active development, this documentation is</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="do">## specific to the following version of R, sesame, sesameData and ExperimentHub:</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">sesame_checkVersion</span>()</span></code></pre></div>
<pre><code>## SeSAMe requires matched versions of R, sesame, sesameData and ExperimentHub.
## Here is the current versions installed:
## R: 4.5.1
## Bioconductor: 3.22
## sesame: 1.29.0
## sesameData: 1.27.1
## ExperimentHub: 3.1.0</code></pre>
<p>We recommend updating your R, ExperimentHub, sesameData and sesame to
use this documentation consistently. If you have installed directly from
github, please make sure the compatible ExperimentHub is installed.</p>
<p>If you use a previous version, please checkout the vignette that
corresponds to the right version here <a href="https://zhou-lab.github.io/sesame/dev/supplemental.html#Versions">https://zhou-lab.github.io/sesame/dev/supplemental.html#Versions</a></p>
<p><strong>CRITICAL:</strong> After a new installation, one must cache
the associated annotation data using the following command. This needs
to be done only once per SeSAMe installation/update. Caching data to
local guarantees proper data retrieval and saves internet traffic.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">sesameDataCache</span>()</span></code></pre></div>
<p>This function caches the needed SeSAMe annotations. SeSAMe annotation
data is managed by the <a href="https://tinyurl.com/58ny3rrt">sesameData</a> package which uses
the <a href="https://tinyurl.com/2p873tez">ExperimentHub</a>
infrastructure. You can find the location of the cached annotation data
on your local computer using:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>tools<span class="sc">::</span><span class="fu">R_user_dir</span>(<span class="st">&quot;ExperimentHub&quot;</span>, <span class="at">which=</span><span class="st">&quot;cache&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;/github/home/.cache/R/ExperimentHub&quot;</code></pre>
</div>
<div id="the-opensesame-pipeline" class="section level1">
<h1>The openSesame Pipeline</h1>
<p>The <code>openSesame</code> function provides end-to-end processing
that converts IDATs to DNA methylation level (aka <span class="math inline">\(\beta\)</span> value) matrices in R. The function
input <code>path_to_idats</code> can be one of the following input:</p>
<ul>
<li>A path to a directory where the IDAT files will be recursively
search</li>
<li>Specific path(s) of IDAT file prefix, one for each IDAT pair</li>
<li>One or a list of SigDF objects</li>
</ul>
<p>The following code uses a directory that contains built-in two HM27
IDAT pairs to demonstrates the use of <code>openSesame</code>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">openSesame</span>(<span class="st">&quot;path_to_idats&quot;</span>, <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="dv">2</span>))</span></code></pre></div>
<p>The <code>BPPARAM=</code> option is from the
<code>BiocParallel</code> package and controls parallel processing (in
this case, we are using two cores). Under the hood, the function
performs a series of tasks including: searching IDAT files from the
directory (the <code>searchIDATprefixes</code> function), reading IDAT
data in as <code>SigDF</code> objects (the <code>readIDATpair</code>
function), preprocessing the signals (the <code>prepSesame</code>
function), and finally converting them to DNA methylation levels (<span class="math inline">\(\beta\)</span> values, the <code>getBetas</code>
function). Alternatively, one can run the following command to get the
same results, while gaining more refined control:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">##  The above openSesame call is equivalent to:</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">do.call</span>(cbind, BiocParallel<span class="sc">::</span><span class="fu">bplapply</span>(</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>    <span class="fu">searchIDATprefixes</span>(idat_dir), <span class="cf">function</span>(pfx) {</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>        <span class="fu">getBetas</span>(<span class="fu">prepSesame</span>(<span class="fu">readIDATpair</span>(pfx), <span class="st">&quot;QCDPB&quot;</span>))</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>}, <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="dv">2</span>)))</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">## or even more explicitly (if one needs to control argument passed</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">## to a specific preprocessing function)</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">do.call</span>(cbind, BiocParallel<span class="sc">::</span><span class="fu">bplapply</span>(</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>    <span class="fu">searchIDATprefixes</span>(idat_dir), <span class="cf">function</span>(pfx) {</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>        <span class="fu">getBetas</span>(<span class="fu">noob</span>(<span class="fu">pOOBAH</span>(<span class="fu">dyeBiasNL</span>(<span class="fu">inferInfiniumIChannel</span>(<span class="fu">qualityMask</span>(</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>            <span class="fu">readIDATpair</span>(pfx)))))))</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>}, <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="dv">2</span>)))</span></code></pre></div>
<p>The <code>openSesame</code> function is highly customizable. The
<code>prep=</code> argument is the same argument one gives to the
<code>prepSesame</code> function (see <a href="#dataprep">Data
Preprocessing</a> for detail) which <code>openSesame</code> calls
internally. The argument uniquely specifies a preprocessing procedure.
The <code>func=</code> option specifies the signal extraction function.
It can be either be <code>getBetas</code> (DNA methylation) or
<code>getAFs</code> (allele frequencies of SNP probes) or NULL (returns
<code>SigDF</code>). The <code>manifest=</code> option allows one to
provide an array manifest when handling data from platform not supported
natively. Finally, the <code>BPPARAM=</code> argument is the same
argument taken by <code>BiocParallel::bplapply</code> to allow parallel
processing. See <a href="https://zhou-lab.github.io/sesame/v1.16/supplemental.html#openSesame">Supplemental
Vignette</a> for details of these component functions of openSesame.</p>
<p>The output of <code>openSesame</code> can also be customized. It can
either be beta values, which are the end DNA methylation readings, as
shown above. It can also be a list of <code>SigDF</code>s which stores
the signal intensities and can be further put back to openSesame for
more processing. The <code>openSesame(func=)</code> argument specifies
whether the output is a SigDF list or beta values. The following shows
some usage:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">openSesame</span>(idat_dir, <span class="at">func =</span> getBetas) <span class="co"># getBetas is the default</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>sdfs <span class="ot">=</span> <span class="fu">openSesame</span>(idat_dir, <span class="at">func =</span> <span class="cn">NULL</span>) <span class="co"># return SigDF list</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>allele_freqs <span class="ot">=</span> <span class="fu">openSesame</span>(idat_dir, <span class="at">func =</span> getAFs) <span class="co"># SNP allele frequencies</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>sdfs <span class="ot">=</span> <span class="fu">openSesame</span>(sdfs, <span class="at">prep =</span> <span class="st">&quot;Q&quot;</span>, <span class="at">func =</span> <span class="cn">NULL</span>)   <span class="co"># take and return SigDFs</span></span></code></pre></div>
<p>One can also generate the detection p-values (e.g., for GEO upload)
by feeding openSesame with <code>func = pOOBAH</code> (or other
detection p-value calculators).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>pvals <span class="ot">=</span> <span class="fu">openSesame</span>(idat_dir, <span class="at">func =</span> pOOBAH, <span class="at">return.pval=</span><span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="dataprep" class="section level1">
<h1>Data Preprocessing</h1>
<p>The <code>prep=</code> argument instructs the <code>openSesame</code>
function to call the <code>prepSesame</code> function to preprocess
signal intensity under the hood. This can be skipped by using
<code>prep=&quot;&quot;</code>. The <code>prepSesame</code> function takes a
single <code>SigDF</code> as input and returns a processed
<code>SigDF</code>. When <code>prep=</code> is non-empty, it selects the
preprocessing functions (see <a href="#prep">Preprocessing Function
Code</a>) and specifies the order of their execution. For example,</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>sdf <span class="ot">=</span> <span class="fu">sesameDataGet</span>(<span class="st">&#39;EPIC.1.SigDF&#39;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>sdf_preped <span class="ot">=</span> <span class="fu">openSesame</span>(sdf, <span class="at">prep=</span><span class="st">&quot;DB&quot;</span>, <span class="at">func=</span><span class="cn">NULL</span>)</span></code></pre></div>
<p>performs dye bias correction (<code>D</code>) followed by background
subtraction (<code>B</code>). In other words,
<code>prepSesame(sdf, &quot;DB&quot;)</code> is equivalent to
<code>noob(dyeBiasNL(sdf))</code>. All the preprocessing functions take
a <code>SigDF</code> as input and return an updated <code>SigDF</code>.
Therefore, these functions can be chained together. The choice of
preprocessing functions and the order of their chaining is important
(see <a href="https://zhou-lab.github.io/sesame/v1.16/supplemental.html#prepfuns">Supplemental
Vignette</a>) for detailed discussions of these functions). The
following table lists the best preprocessing strategy based on our
experience.</p>
<table>
<caption>Recommended Preprocessing</caption>
<thead>
<tr class="header">
<th align="left">Platform</th>
<th align="left">Sample Organism</th>
<th align="left">Prep Code</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">EPICv2/EPIC/HM450</td>
<td align="left">human</td>
<td align="left">QCDPB</td>
</tr>
<tr class="even">
<td align="left">EPICv2/EPIC/HM450</td>
<td align="left">non-human organism</td>
<td align="left">SQCDPB</td>
</tr>
<tr class="odd">
<td align="left">MM285</td>
<td align="left">mouse</td>
<td align="left">TQCDPB</td>
</tr>
<tr class="even">
<td align="left">MM285</td>
<td align="left">non-mouse organism</td>
<td align="left">SQCDPB</td>
</tr>
<tr class="odd">
<td align="left">Mammal40</td>
<td align="left">human</td>
<td align="left">HCDPB</td>
</tr>
<tr class="even">
<td align="left">Mammal40</td>
<td align="left">non-human organism</td>
<td align="left">SHCDPB</td>
</tr>
</tbody>
</table>
<p>The optimal strategy of preprocessing depends on:</p>
<ol style="list-style-type: decimal">
<li><p><strong>The array platform</strong>. For example, certain array
platforms (e.g., the Mammal40) do not have enough Infinium-I probes for
background estimation and dye bias correction, therefore background
subtraction (where the out-of-band signals are from) might not work most
optimally;</p></li>
<li><p><strong>The expected sample property</strong>. For example, some
samples have the signature bimodal distribution of methylation of most
mammalian cells. Others may undergo global loss of methylation (germ
cells, tumors etc). Other important factors include high-input vs
low-input, tumor vs normal, somatic vs germ cells, human vs model
organisms, mouse strains etc. Some platforms (e.g., Mammal40 and MM285)
are designed for multiple species and strains. Therefore <code>S</code>
and <code>T</code> would be important when those arrays are used on
non-reference organisms (see <a href="nonhuman.html">Working with
Nonhuman Arrays</a>).</p></li>
</ol>
</div>
<div id="prep" class="section level1">
<h1>Preprocessing Function Code</h1>
<p>The <code>prepSesameList</code> function lists all the available
codes and the associated preprocessing functions.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">prepSesameList</span>()</span></code></pre></div>
<pre><code>##    code                  func
## 1     0             resetMask
## 2     Q           qualityMask
## 3     G       prefixMaskButCG
## 4     H        prefixMaskButC
## 5     C inferInfiniumIChannel
## 6     D             dyeBiasNL
## 7     E              dyeBiasL
## 8     P                pOOBAH
## 9     I                 ELBAR
## 10    B                  noob
## 11    U                 scrub
## 12    S          inferSpecies
## 13    T           inferStrain
## 14    M           matchDesign
##                                           description
## 1                             Reset mask to all FALSE
## 2                          Mask probes of poor design
## 3                             Mask all but cg- probes
## 4                      Mask all but cg- and ch-probes
## 5                 Infer channel for Infinium-I probes
## 6                    Dye bias correction (non-linear)
## 7                        Dye bias correction (linear)
## 8                 Detection p-value masking using oob
## 9                  Mask background-dominated readings
## 10                   Background subtraction using oob
## 11 More aggressive background subtraction using scrub
## 12                          Set species-specific mask
## 13                   Set strain-specific mask (mouse)
## 14                Match Inf-I/II in beta distribution</code></pre>
<p>Here are some consideration when determining the preprocessing order.
Species <code>(S)</code> and strain <code>(T)</code> inference resets
the mask and color channels based on probe alignment and presence of
genetic variants. Therefore when they are used, they need to be called
first. <code>Q</code> masks non-uniquely mapped probes which may inflate
the out-of-band signal for background estimation. Therefore
<code>Q</code> should be used before detection p-value calculation
(<code>P</code>) and background subtraction (<code>B</code>) when
necessary. Channel inference (<code>C</code>) and dye bias correction
(<code>D</code>) should take place early since dye bias effect is
global. <code>C</code> should be placed before <code>D</code> because
dye bias correction uses in-band signal the identification of which
relies on correct channel designation. Detection p-value
(<code>P</code>) should happen before background subtraction
(<code>B</code>) since background subtraction modifies signal and may
affect out-of-band signal assumption used in <code>P</code>. Lastly,
functions that explicitly normalizes <span class="math inline">\(\beta\)</span> value distribution (<code>M</code>)
should happen last if they even need to be used.</p>
<p>See <a href="https://zhou-lab.github.io/sesame/v1.16/supplemental.html#prepfuns">Supplemental
Vignette</a> for details of preprocessing functions.</p>
</div>
<div id="lift-over-across-platforms" class="section level1">
<h1>Lift over across platforms</h1>
<p>To allow data integration across platforms, one can harmonize beta
values, probe ID list and even signal intensities, to a different
platform using the <code>mLiftOver</code> (methylation Lift Over)
utility. The following examples illustrates these corresponding
usages.</p>
<div id="project-probe-ids" class="section level2">
<h2>Project probe IDs</h2>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>cg_msa <span class="ot">=</span> <span class="fu">names</span>(<span class="fu">sesameData_getManifestGRanges</span>(<span class="st">&quot;MSA&quot;</span>))</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="do">## only mappable probes, return mapping from MSA to HM450</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">mLiftOver</span>(cg_msa, <span class="st">&quot;HM450&quot;</span>))</span></code></pre></div>
<pre><code>## cg13869341_BC11 cg12045430_BC11 cg18231760_BC21 cg17866181_BC21 cg08477687_BC21 
##    &quot;cg13869341&quot;    &quot;cg12045430&quot;    &quot;cg18231760&quot;    &quot;cg17866181&quot;    &quot;cg08477687&quot; 
## cg23917638_TC21 
##    &quot;cg23917638&quot;</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>cg_hm450 <span class="ot">=</span> <span class="fu">names</span>(<span class="fu">sesameData_getManifestGRanges</span>(<span class="st">&quot;HM450&quot;</span>))</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>cg_hm450 <span class="ot">=</span> <span class="fu">grep</span>(<span class="st">&quot;cg&quot;</span>, cg_hm450, <span class="at">value=</span><span class="cn">TRUE</span>)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="do">## only mappable probes, return mapping from HM450 to EPICv2</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">mLiftOver</span>(cg_hm450, <span class="st">&quot;EPICv2&quot;</span>))</span></code></pre></div>
<pre><code>##        cg00381604        cg21870274        cg06402284        cg23803172 
## &quot;cg00381604_BC11&quot; &quot;cg21870274_BC21&quot; &quot;cg06402284_TC21&quot; &quot;cg23803172_BC11&quot; 
##        cg20788133        cg08258224 
## &quot;cg20788133_BC21&quot; &quot;cg08258224_TC21&quot;</code></pre>
</div>
<div id="project-beta-values" class="section level2">
<h2>Project beta values</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">openSesame</span>(<span class="fu">sesameDataGet</span>(<span class="st">&quot;EPICv2.8.SigDF&quot;</span>)[[<span class="dv">1</span>]])</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>betas_epic <span class="ot">=</span> <span class="fu">mLiftOver</span>(betas, <span class="st">&quot;EPIC&quot;</span>, <span class="at">impute=</span><span class="cn">FALSE</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="fu">length</span>(betas_epic)     <span class="co"># EPIC platform dimension</span></span></code></pre></div>
<pre><code>## [1] 866553</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(betas_epic)) <span class="co"># NA values are present</span></span></code></pre></div>
<pre><code>## [1] 152102</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>betas_epic <span class="ot">=</span> <span class="fu">imputeBetas</span>(betas_epic)</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="fu">length</span>(betas_epic)     <span class="co"># EPIC platform dimension</span></span></code></pre></div>
<pre><code>## [1] 866553</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(betas_epic)) <span class="co"># expect 0 NA after imputation</span></span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="do">## use empirical evidence in mLiftOver</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>mapping <span class="ot">=</span> <span class="fu">sesameDataGet</span>(<span class="st">&quot;liftOver.EPICv2ToEPIC&quot;</span>)</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>betas_matrix <span class="ot">=</span> <span class="fu">openSesame</span>(<span class="fu">sesameDataGet</span>(<span class="st">&quot;EPICv2.8.SigDF&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])</span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">mLiftOver</span>(betas_matrix, <span class="st">&quot;EPIC&quot;</span>, <span class="at">mapping =</span> mapping))</span></code></pre></div>
<pre><code>## [1] 539513      2</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="do">## compare to without using empirical evidence</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">mLiftOver</span>(betas_matrix, <span class="st">&quot;EPIC&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 866553      2</code></pre>
</div>
<div id="project-signal-sigdfs" class="section level2">
<h2>Project signal SigDFs</h2>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a>sdf <span class="ot">=</span> <span class="fu">sesameDataGet</span>(<span class="st">&quot;EPICv2.8.SigDF&quot;</span>)[[<span class="st">&quot;GM12878_206909630042_R08C01&quot;</span>]]</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">mLiftOver</span>(sdf, <span class="st">&quot;EPICv2&quot;</span>)) <span class="co"># EPICv2 platform dimension</span></span></code></pre></div>
<pre><code>## [1] 937690      7</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">mLiftOver</span>(sdf, <span class="st">&quot;EPIC&quot;</span>))   <span class="co"># EPIC platform dimension</span></span></code></pre></div>
<pre><code>## [1] 866553      7</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">mLiftOver</span>(sdf, <span class="st">&quot;HM450&quot;</span>))  <span class="co"># HM450 platform dimension</span></span></code></pre></div>
<pre><code>## [1] 486427      7</code></pre>
</div>
</div>
<div id="collapse-to-cg-prefixes" class="section level1">
<h1>Collapse to cg prefixes</h1>
<p>The more recent Infinium arrays (the mouse, EPICv2, and MSA arrays)
have suffixes to uniquely identify the probe design. If one prefers to
collapse methylation readings to the cg prefix so they are comparable to
the previous array generations, this can be done with
<code>getBetas(..., collapseToPfx = TRUE)</code>. The same argument can
be passed to
<code>openSesame(..., func=getBetas, collapseToPfx=TRUE)</code></p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">getBetas</span>(sdf_from_EPICv2, <span class="at">collapseToPfx =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a><span class="do">## or</span></span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">openSesame</span>(<span class="st">&quot;path_to_idats&quot;</span>, <span class="at">collapseToPfx =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a><span class="do">## by default the method for collapsing is to make means</span></span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">openSesame</span>(<span class="st">&quot;path_to_idats&quot;</span>, <span class="at">collapseToPfx =</span> <span class="cn">TRUE</span>, <span class="at">collapseMethod =</span> <span class="st">&quot;mean&quot;</span>)</span>
<span id="cb36-6"><a href="#cb36-6" tabindex="-1"></a><span class="do">## one can also switch to min detection p-value</span></span>
<span id="cb36-7"><a href="#cb36-7" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">openSesame</span>(<span class="st">&quot;path_to_idats&quot;</span>, <span class="at">collapseToPfx =</span> <span class="cn">TRUE</span>, <span class="at">collapseMethod =</span> <span class="st">&quot;minPval&quot;</span>)</span></code></pre></div>
<p>One may also use the <code>betasCollapseToPfx</code> function to
generate the traditional cg number readings.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">betasCollapseToPfx</span>(betas, <span class="at">BPPARAM=</span>BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="dv">2</span>))</span></code></pre></div>
<p>Please find more info at the Supplemental documentation at the
following [link] (<a href="https://zhou-lab.github.io/sesame/dev/supplemental.html#Replicate_Probes" class="uri">https://zhou-lab.github.io/sesame/dev/supplemental.html#Replicate_Probes</a>)</p>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.5.1 (2025-06-13)
## Platform: x86_64-pc-linux-gnu
## Running under: Ubuntu 24.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Etc/UTC
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] wheatmap_0.2.0              ggplot2_4.0.0              
##  [3] tidyr_1.3.1                 dplyr_1.1.4                
##  [5] SummarizedExperiment_1.41.0 Biobase_2.71.0             
##  [7] GenomicRanges_1.63.0        Seqinfo_1.1.0              
##  [9] IRanges_2.45.0              S4Vectors_0.49.0           
## [11] MatrixGenerics_1.23.0       matrixStats_1.5.0          
## [13] knitr_1.50                  sesame_1.29.0              
## [15] sesameData_1.27.1           ExperimentHub_3.1.0        
## [17] AnnotationHub_4.1.0         BiocFileCache_3.1.0        
## [19] dbplyr_2.5.1                BiocGenerics_0.57.0        
## [21] generics_0.1.4             
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.2.1      farver_2.1.2          blob_1.2.4           
##  [4] filelock_1.0.3        Biostrings_2.79.1     S7_0.2.0             
##  [7] fastmap_1.2.0         digest_0.6.37         lifecycle_1.0.4      
## [10] KEGGREST_1.51.0       RSQLite_2.4.3         magrittr_2.0.4       
## [13] compiler_4.5.1        rlang_1.1.6           sass_0.4.10          
## [16] tools_4.5.1           utf8_1.2.6            yaml_2.3.10          
## [19] labeling_0.4.3        S4Arrays_1.11.0       bit_4.6.0            
## [22] curl_7.0.0            DelayedArray_0.37.0   plyr_1.8.9           
## [25] RColorBrewer_1.1-3    abind_1.4-8           BiocParallel_1.45.0  
## [28] withr_3.0.2           purrr_1.1.0           sys_3.4.3            
## [31] grid_4.5.1            preprocessCore_1.73.0 colorspace_2.1-2     
## [34] MASS_7.3-65           scales_1.4.0          dichromat_2.0-0.1    
## [37] cli_3.6.5             rmarkdown_2.30        crayon_1.5.3         
## [40] reshape2_1.4.4        httr_1.4.7            tzdb_0.5.0           
## [43] DBI_1.2.3             cachem_1.1.0          stringr_1.5.2        
## [46] splines_4.5.1         parallel_4.5.1        AnnotationDbi_1.73.0 
## [49] BiocManager_1.30.26   XVector_0.51.0        vctrs_0.6.5          
## [52] Matrix_1.7-4          jsonlite_2.0.0        hms_1.1.4            
## [55] bit64_4.6.0-1         maketools_1.3.2       jquerylib_0.1.4      
## [58] glue_1.8.0            codetools_0.2-20      stringi_1.8.7        
## [61] gtable_0.3.6          BiocVersion_3.23.1    tibble_3.3.0         
## [64] pillar_1.11.1         rappdirs_0.3.3        htmltools_0.5.8.1    
## [67] R6_2.6.1              httr2_1.2.1           evaluate_1.0.5       
## [70] lattice_0.22-7        readr_2.1.5           png_0.1-8            
## [73] memoise_2.0.1         BiocStyle_2.39.0      bslib_0.9.0          
## [76] Rcpp_1.1.0            nlme_3.1-168          SparseArray_1.11.1   
## [79] mgcv_1.9-1            xfun_0.54             buildtools_1.0.0     
## [82] pkgconfig_2.0.3</code></pre>
</div></div>
            <div class="d-none d-xl-block col-xl-2 bd-toc">
               <ul class="section-nav">
                  <li class="toc-entry"><ul>
<li><a href="#preparation" id="toc-preparation">Preparation</a></li>
<li><a href="#the-opensesame-pipeline" id="toc-the-opensesame-pipeline">The openSesame Pipeline</a></li>
<li><a href="#dataprep" id="toc-dataprep">Data Preprocessing</a></li>
<li><a href="#prep" id="toc-prep">Preprocessing Function Code</a></li>
<li><a href="#lift-over-across-platforms" id="toc-lift-over-across-platforms">Lift over across platforms</a>
<ul>
<li><a href="#project-probe-ids" id="toc-project-probe-ids">Project
probe IDs</a></li>
<li><a href="#project-beta-values" id="toc-project-beta-values">Project
beta values</a></li>
<li><a href="#project-signal-sigdfs" id="toc-project-signal-sigdfs">Project signal SigDFs</a></li>
</ul></li>
<li><a href="#collapse-to-cg-prefixes" id="toc-collapse-to-cg-prefixes">Collapse to cg prefixes</a></li>
<li><a href="#session-info" id="toc-session-info">Session Info</a></li>
</ul></li>
               </ul>
            </div>
         </div>
               </div>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.slim.js" data-external="1"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.6.0/js/bootstrap.min.js" data-external="1"></script>
      <script src="https://r-universe.dev/static/article.js" data-external="1"></script>

      <!-- Mathjax -->
      <!-- <script src="https://polyfill.io/v3/polyfill.min.js?features=es6" data-external="1"></script> -->
      <script id="MathJax-script" async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/4.0.0/tex-mml-chtml.min.js" data-external="1"></script>
   <script defer src="https://static.cloudflareinsights.com/beacon.min.js/vcd15cbe7772f49c399c6a5babf22c1241717689176015" integrity="sha512-ZpsOmlRQV6y907TI0dKBHq9Md29nnaEIPlkf84rnaERnq6zvWvPUqr2ft8M1aS28oN72PdrCzSjY4U6VaAw1EQ==" data-cf-beacon='{"version":"2024.11.0","token":"6000532f154d4405a1c967fdafb41d60","r":1,"server_timing":{"name":{"cfCacheStatus":true,"cfEdge":true,"cfExtPri":true,"cfL4":true,"cfOrigin":true,"cfSpeedBrain":true},"location_startswith":null}}' crossorigin="anonymous"></script>
</body>
</html>
