<!DOCTYPE html>
<html>
  <head>
    <title>Package 'variancePartition' reference manual</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

    <!-- styling for math. katex.min.css expects fonts in same dir so use CDN -->
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.9/katex.min.css">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/prism/1.29.0/themes/prism.css">
    <link rel="stylesheet" type="text/css" href="https://r-universe.dev/static/R.css">

    <!-- for (future) customizations -->
    <script src="https://r-universe.dev/static/manual.js"></script>
    <link rel="stylesheet" type="text/css" href="https://r-universe.dev/static/manual.css">

  </head>

  <body class="postdoc macintosh">
    <h1 class="manual-title">Package 'variancePartition'</h1>
    <table class="description-table">
      <tr>
<th>Title:</th>
<td class="description-title">Quantify and interpret drivers of variation in multilevel gene
expression experiments</td>
</tr>
      <tr>
<th>Description:</th>
<td class="description-description">Quantify and interpret multiple sources of biological and
        technical variation in gene expression experiments. Uses a
        linear mixed model to quantify variation in gene expression
        attributable to individual, tissue, time point, or technical
        variables.  Includes dream differential expression analysis for
        repeated measures.</td>
</tr>
      <tr>
<th>Authors:</th>
<td class="description-author"><span>Gabriel Hoffman [aut, cre] (ORCID:
    &lt;https://orcid.org/0000-0002-0957-0224&gt;)</span></td>
</tr>
      <tr>
<th>Maintainer:</th>
<td class="description-maintainer">Gabriel E. Hoffman &lt;<a href="/cdn-cgi/l/email-protection" class="__cf_email__" data-cfemail="610600031308040d4f090e07070c000f210c12120c4f040514">[email&#160;protected]</a>&gt;</td>
</tr>
      <tr>
<th>License:</th>
<td class="description-license">GPL-2</td>
</tr>
      <tr>
<th>Version:</th>
<td class="description-version">1.41.0</td>
</tr>
      <tr>
<th>Built:</th>
<td class="description-date">2025-10-30 14:42:19 UTC</td>
</tr>
      <tr>
<th>Source:</th>
<td class="description-source">https://github.com/bioc/variancePartition</td>
</tr>
    </table>

    <a href="#help-index" style="color:black;"><h2 id="help-index">Help Index</h2></a>
    <ul id="help-index-list">
<li class="help-index-item"><a href="#applyQualityWeights">Apply pre-specified sample weights</a></li>
<li class="help-index-item"><a href="#as.data.frame.varPartResults">Convert to data.frame</a></li>
<li class="help-index-item"><a href="#as.matrix-varPartResults-method">Convert to matrix</a></li>
<li class="help-index-item"><a href="#augmentPriorCount">Augment observed read counts with prior count</a></li>
<li class="help-index-item"><a href="#BIC.MArrayLM">BIC from model fit</a></li>
<li class="help-index-item"><a href="#BIC.MArrayLM2">BIC from model fit</a></li>
<li class="help-index-item"><a href="#calcVarPart-method">Compute variance statistics</a></li>
<li class="help-index-item"><a href="#canCorPairs">canCorPairs</a></li>
<li class="help-index-item"><a href="#classifyTestsF">Multiple Testing Genewise Across Contrasts</a></li>
<li class="help-index-item"><a href="#classifyTestsF-MArrayLM2-method">Multiple Testing Genewise Across Contrasts</a></li>
<li class="help-index-item"><a href="#colinearityScore">Collinearity score</a></li>
<li class="help-index-item"><a href="#deviation-method">Deviation from expectation for each observation</a></li>
<li class="help-index-item"><a href="#diffVar-method">Test differential variance</a></li>
<li class="help-index-item"><a href="#dream-method">Differential expression with linear mixed model</a></li>
<li class="help-index-item"><a href="#dscchisq">Scaled chi-square</a></li>
<li class="help-index-item"><a href="#eBayes">eBayes generic for for MArrayLM and MArrayLM2</a></li>
<li class="help-index-item"><a href="#eBayes-MArrayLM-method">eBayes for MArrayLM</a></li>
<li class="help-index-item"><a href="#eBayes-method">eBayes for MArrayLM2</a></li>
<li class="help-index-item"><a href="#ESS-method">Effective sample size</a></li>
<li class="help-index-item"><a href="#extractVarPart">Extract variance statistics</a></li>
<li class="help-index-item"><a href="#fitExtractVarPartModel-method">Fit linear (mixed) model, report variance fractions</a></li>
<li class="help-index-item"><a href="#fitVarPartModel-method">Fit linear (mixed) model</a></li>
<li class="help-index-item"><a href="#get_prediction-method">Compute predicted value of formula for linear (mixed) model</a></li>
<li class="help-index-item"><a href="#getContrast-method">Extract contrast matrix for linear mixed model</a></li>
<li class="help-index-item"><a href="#getTreat-method">Test if coefficient is different from a specified value</a></li>
<li class="help-index-item"><a href="#ggColorHue">Default colors for ggplot</a></li>
<li class="help-index-item"><a href="#hatvalues-method">Compute hatvalues</a></li>
<li class="help-index-item"><a href="#isRunableFormula">Test if formula is full rank on this dataset</a></li>
<li class="help-index-item"><a href="#logLik.MArrayLM">Log-likelihood from model fit</a></li>
<li class="help-index-item"><a href="#logLik.MArrayLM2">Log-likelihood from model fit</a></li>
<li class="help-index-item"><a href="#makeContrastsDream">Construct Matrix of Custom Contrasts</a></li>
<li class="help-index-item"><a href="#MArrayLM2-class">Class MArrayLM2</a></li>
<li class="help-index-item"><a href="#mvTest-method">Multivariate tests on results from dream()</a></li>
<li class="help-index-item"><a href="#mvTest_input-class">Class mvTest_input</a></li>
<li class="help-index-item"><a href="#plotCompareP-method">Compare p-values from two analyses</a></li>
<li class="help-index-item"><a href="#plotContrasts">Plot representation of contrast matrix</a></li>
<li class="help-index-item"><a href="#plotCorrMatrix">plotCorrMatrix</a></li>
<li class="help-index-item"><a href="#plotCorrStructure">plotCorrStructure</a></li>
<li class="help-index-item"><a href="#plotPercentBars-method">Bar plot of gene fractions</a></li>
<li class="help-index-item"><a href="#plotStratify">plotStratify</a></li>
<li class="help-index-item"><a href="#plotStratifyBy">plotStratifyBy</a></li>
<li class="help-index-item"><a href="#plotVarianceEstimates">Plot Variance Estimates</a></li>
<li class="help-index-item"><a href="#plotVarPart-method">Violin plot of variance fractions</a></li>
<li class="help-index-item"><a href="#rdf">Residual degrees of freedom</a></li>
<li class="help-index-item"><a href="#rdf_from_matrices">Fast approximate residual degrees of freedom</a></li>
<li class="help-index-item"><a href="#rdf.merMod">Approximate residual degrees of freedom</a></li>
<li class="help-index-item"><a href="#reOnly">Adapted from lme4:::reOnly</a></li>
<li class="help-index-item"><a href="#residuals-MArrayLM-method">residuals for MArrayLM</a></li>
<li class="help-index-item"><a href="#residuals-MArrayLM2-method">residuals for MArrayLM2</a></li>
<li class="help-index-item"><a href="#residuals-VarParFitList-method">Residuals from model fit</a></li>
<li class="help-index-item"><a href="#residuals.MArrayLM2">Residuals for result of dream</a></li>
<li class="help-index-item"><a href="#shrinkageMetric">Shrinkage metric for eBayes</a></li>
<li class="help-index-item"><a href="#sortCols-method">Sort variance partition statistics</a></li>
<li class="help-index-item"><a href="#toptable-method">Table of Top Genes from Linear Model Fit</a></li>
<li class="help-index-item"><a href="#VarParCIList-class">Class VarParCIList</a></li>
<li class="help-index-item"><a href="#VarParFitList-class">Class VarParFitList</a></li>
<li class="help-index-item"><a href="#varParFrac-class">Class varParFrac</a></li>
<li class="help-index-item"><a href="#varPartConfInf">Linear mixed model confidence intervals</a></li>
<li class="help-index-item"><a href="#varPartData">Simulation dataset for examples</a></li>
<li class="help-index-item"><a href="#varPartDEdata">A simulated dataset of gene counts</a></li>
<li class="help-index-item"><a href="#varPartResults-class">Class varPartResults</a></li>
<li class="help-index-item"><a href="#vcov-MArrayLM-method">Co-variance matrix for dream() fit</a></li>
<li class="help-index-item"><a href="#vcov-MArrayLM2-method">Co-variance matrix for dream() fit</a></li>
<li class="help-index-item"><a href="#vcovSqrt-method">Sqrt of co-variance matrix for dream() fit</a></li>
<li class="help-index-item"><a href="#voomWithDreamWeights">Transform RNA-Seq Data Ready for Linear Mixed Modelling with dream()</a></li>
</ul>

    <hr>
    <div class="manual-pages-content">
<div class="container manual-page" id="applyQualityWeights"><div class="page-main">
<a href="#applyQualityWeights" class="help-page-title"><h2>Apply pre-specified sample weights</h2></a>

<h3>Description</h3>

<p>Apply pre-specified sample weights by scaling existing precision weights
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">applyQualityWeights(vobj, weights)
</code><code class="language-r">applyQualityWeights<span class="token punctuation">(</span>vobj<span class="token punctuation">,</span> weights<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> from <code>voom</code> or <code>voomWithDreamWeights</code>.</p>
</td>
</tr>
<tr>
<td><code id="weights">weights</code></td>
<td>
<p>sample level weights</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Apply pre-specified sample-level weights to the existing precision weights estimated from the data.  While the <code>limma::voomWithQualityWeights</code> function of Lui et al. (2015) estimates the sample-level weights from <code>voom</code> fit, here the weights are fixed beforehand.
</p>


<h3>References</h3>


<p>Liu R, Holik AZ, Su S, Jansz N, Chen K, Leong HS, Blewitt ME, Asselin-Labat M, Smyth GK, Ritchie ME (2015).
“Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses.”
<em>Nucleic acids research</em>, <b>43</b>(15), e97–e97.

</p>


<h3>See Also</h3>

<p><code>limma::voomWithQualityWeights</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="as.data.frame.varPartResults"><div class="page-main">
<a href="#as.data.frame.varPartResults" class="help-page-title"><h2>Convert to data.frame</h2></a>

<h3>Description</h3>

<p>Convert varPartResults to data.frame
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S3 method for class 'varPartResults'
as.data.frame(x, row.names = NULL, optional = FALSE, ...)
</code><code class="language-r"><span class="token comment">## S3 method for class 'varPartResults'</span>
as.data.frame<span class="token punctuation">(</span>x<span class="token punctuation">,</span> row.names <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span> optional <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="x">x</code></td>
<td>
<p>varPartResults</p>
</td>
</tr>
<tr>
<td><code id="row.names">row.names</code></td>
<td>
<p>pass thru to generic</p>
</td>
</tr>
<tr>
<td><code id="optional">optional</code></td>
<td>
<p>pass thru to generic</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments.</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>data.frame
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># load library
# library(variancePartition)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Fit model
varPart &lt;- fitExtractVarPartModel(geneExpr[1:5, ], form, info)

# convert to matrix
as.data.frame(varPart)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Fit model</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">5</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># convert to matrix</span>
as.data.frame<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="as.matrix-varPartResults-method"><div class="page-main">
<a href="#as.matrix-varPartResults-method" class="help-page-title"><h2>Convert to matrix</h2></a>

<h3>Description</h3>

<p>Convert varPartResults to matrix
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'varPartResults'
as.matrix(x, ...)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'varPartResults'</span>
as.matrix<span class="token punctuation">(</span>x<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="x">x</code></td>
<td>
<p>varPartResults</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments.</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>matrix
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># load library
# library(variancePartition)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Fit model
varPart &lt;- fitExtractVarPartModel(geneExpr[1:5, ], form, info)

# convert to matrix
as.matrix(varPart)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Fit model</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">5</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># convert to matrix</span>
as.matrix<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="augmentPriorCount"><div class="page-main">
<a href="#augmentPriorCount" class="help-page-title"><h2>Augment observed read counts with prior count</h2></a>

<h3>Description</h3>

<p>Augment observed read counts with prior count since log of zero counts is undefined.  The prior count added to each sample is scaled so that no variance is introduced
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">augmentPriorCount(
  counts,
  lib.size = colSums2(counts),
  prior.count = 0.5,
  scaledByLib = FALSE
)
</code><code class="language-r">augmentPriorCount<span class="token punctuation">(</span>
  counts<span class="token punctuation">,</span>
  lib.size <span class="token operator">=</span> colSums2<span class="token punctuation">(</span>counts<span class="token punctuation">)</span><span class="token punctuation">,</span>
  prior.count <span class="token operator">=</span> <span class="token number">0.5</span><span class="token punctuation">,</span>
  scaledByLib <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="counts">counts</code></td>
<td>
<p>matrix of read counts with genes as rows and samples as columns</p>
</td>
</tr>
<tr>
<td><code id="lib.size">lib.size</code></td>
<td>
<p>library sizes, the sum of all ready for each sample</p>
</td>
</tr>
<tr>
<td><code id="prior.count">prior.count</code></td>
<td>
<p>average prior count added to each sample.</p>
</td>
</tr>
<tr>
<td><code id="scaledByLib">scaledByLib</code></td>
<td>
<p>if <code>TRUE</code>, scale pseudocount by <code>lib.size</code>.  Else to standard constant pseudocount addition</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Adding prior counts removes the issue of evaluating the log of zero counts, and stabilizes the log transform when counts are very small. However, adding a constant prior count to all samples can introduced an artifact. Consider two samples each with zero counts for a given gene, but one as a library size of 1k and the other of 50k. After applying the prior count values become pc / 1k and pc / 50k. It appears that there is variance in the expression of this gene, even though no counts are observed. This is driven only by variation in the library size, which does not reflect biology. This issue is most problematic for small counts.
</p>
<p>Instead, we make the reasonable assumption that a gene does not have expression variance unless supported sufficiently by counts in the numerator.  Consider adding a different prior count to each sample so that genes with zero counts end up woth zero variance. This corresponds to adding <code>prior.count * lib.size[i] / mean(lib.size)</code> to sample <code>i</code>.  
</p>
<p>This is done in the backend of <code>edgeR::cpm()</code>, but this function allows users to apply it more generally.
</p>


<h3>Value</h3>

<p>matrix with augmented counts
</p>


<h3>See Also</h3>

<p><code>edgeR::cpm()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">library(edgeR)

data(varPartDEdata)

# normalize RNA-seq counts
dge &lt;- DGEList(counts = countMatrix)
dge &lt;- calcNormFactors(dge)

countsAugmented &lt;- augmentPriorCount( dge$counts, dge$samples$lib.size, 1)
</code><code class="language-r">library<span class="token punctuation">(</span>edgeR<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartDEdata<span class="token punctuation">)</span>

<span class="token comment"># normalize RNA-seq counts</span>
dge <span class="token operator">&lt;-</span> DGEList<span class="token punctuation">(</span>counts <span class="token operator">=</span> countMatrix<span class="token punctuation">)</span>
dge <span class="token operator">&lt;-</span> calcNormFactors<span class="token punctuation">(</span>dge<span class="token punctuation">)</span>

countsAugmented <span class="token operator">&lt;-</span> augmentPriorCount<span class="token punctuation">(</span> dge<span class="token operator">$</span>counts<span class="token punctuation">,</span> dge<span class="token operator">$</span>samples<span class="token operator">$</span>lib.size<span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="BIC.MArrayLM"><div class="page-main">
<a href="#BIC.MArrayLM" class="help-page-title"><h2>BIC from model fit</h2></a>

<h3>Description</h3>

<p>BIC from model fit
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S3 method for class 'MArrayLM'
BIC(object, vobj, ...)
</code><code class="language-r"><span class="token comment">## S3 method for class 'MArrayLM'</span>
BIC<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>result of <code>lmFit()</code> or <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> used to fit model</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>See <code>?stats::BIC</code></p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="BIC.MArrayLM2"><div class="page-main">
<a href="#BIC.MArrayLM2" class="help-page-title"><h2>BIC from model fit</h2></a>

<h3>Description</h3>

<p>BIC from model fit using edf
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S3 method for class 'MArrayLM2'
BIC(object, vobj, ...)
</code><code class="language-r"><span class="token comment">## S3 method for class 'MArrayLM2'</span>
BIC<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>result of <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> used to fit model</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>See <code>?stats::BIC</code></p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="calcVarPart-method"><div class="page-main">
<a href="#calcVarPart-method" class="help-page-title"><h2>Compute variance statistics</h2></a>

<h3>Description</h3>

<p>Compute fraction of variation attributable to each variable in regression model.  Also interpretable as the intra-class correlation after correcting for all other variables in the model.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">calcVarPart(fit, returnFractions = TRUE, ...)

## S4 method for signature 'lm'
calcVarPart(fit, returnFractions = TRUE, ...)

## S4 method for signature 'lmerMod'
calcVarPart(fit, returnFractions = TRUE, ...)

## S4 method for signature 'glm'
calcVarPart(fit, returnFractions = TRUE, ...)

## S4 method for signature 'negbin'
calcVarPart(fit, returnFractions = TRUE, ...)

## S4 method for signature 'glmerMod'
calcVarPart(fit, returnFractions = TRUE, ...)
</code><code class="language-r">calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> returnFractions <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'lm'</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> returnFractions <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'lmerMod'</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> returnFractions <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'glm'</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> returnFractions <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'negbin'</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> returnFractions <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'glmerMod'</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> returnFractions <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit from lm() or lmer()</p>
</td>
</tr>
<tr>
<td><code id="returnFractions">returnFractions</code></td>
<td>
<p>default: TRUE.  If TRUE return fractions that sum to 1.  Else return unscaled variance components.</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>additional arguments (not currently used)</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>For linear model, variance fractions are computed based on the sum of squares explained by each component.  For the linear mixed model, the variance fractions are computed by variance component estimates for random effects and sum of squares for fixed effects.
</p>
<p>For a generalized linear model, the variance fraction also includes the contribution of the link function so that fractions are reported on the linear (i.e. link) scale rather than the observed (i.e. response) scale. For linear regression with an identity link, fractions are the same on both scales.  But for logit or probit links, the fractions are not well defined on the observed scale due to the transformation imposed by the link function.
</p>
<p>The variance implied by the link function is the variance of the corresponding distribution:
</p>
<p>logit -&gt; logistic distribution -&gt; variance is <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>π</mi><mfrac><mn>2</mn><mn>3</mn></mfrac></msup></mrow><annotation encoding="application/x-tex">\pi^\frac{2}{3}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.954em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">π</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.954em;"><span style="top:-3.363em;margin-right:0.05em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mopen nulldelimiter sizing reset-size3 size6"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.8443em;"><span style="top:-2.656em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight"><span class="mord mtight">3</span></span></span></span><span style="top:-3.2255em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line mtight" style="border-bottom-width:0.049em;"></span></span><span style="top:-3.384em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.344em;"><span></span></span></span></span></span><span class="mclose nulldelimiter sizing reset-size3 size6"></span></span></span></span></span></span></span></span></span></span></span></span></code>
</p>
<p>probit -&gt; standard normal distribution -&gt; variance is 1
</p>
<p>For the Poisson distribution with rate <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>λ</mi></mrow><annotation encoding="application/x-tex">\lambda</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em;"></span><span class="mord mathnormal">λ</span></span></span></span></code>, the variance is <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><mi>g</mi><mo stretchy="false">(</mo><mn>1</mn><mo>+</mo><mn>1</mn><mi mathvariant="normal">/</mi><mi>λ</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">log(1 + 1/\lambda)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">1/</span><span class="mord mathnormal">λ</span><span class="mclose">)</span></span></span></span></code>.
</p>
<p>For the negative binomial distribution with rate <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>λ</mi></mrow><annotation encoding="application/x-tex">\lambda</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em;"></span><span class="mord mathnormal">λ</span></span></span></span></code> and shape <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>θ</mi></mrow><annotation encoding="application/x-tex">\theta</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">θ</span></span></span></span></code>, the variance is <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><mi>g</mi><mo stretchy="false">(</mo><mn>1</mn><mo>+</mo><mn>1</mn><mi mathvariant="normal">/</mi><mi>λ</mi><mo>+</mo><mn>1</mn><mi mathvariant="normal">/</mi><mi>θ</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">log(1 + 1/\lambda + 1/\theta)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">1/</span><span class="mord mathnormal">λ</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">1/</span><span class="mord mathnormal" style="margin-right:0.02778em;">θ</span><span class="mclose">)</span></span></span></span></code>.
</p>
<p>Variance decomposition is reviewed by Nakagawa and Schielzeth (2012), and expanded to other GLMs by Nakagawa, et al. (2017).  See McKelvey and Zavoina (1975) for early work on applying to GLMs.  Also see DeMaris (2002).
</p>
<p>We note that Nagelkerke's pseudo R^2 evaluates the variance explained by the full model.  Instead, a variance partitioning approach evaluates the variance explained by each term in the model, so that the sum of each systematic plus random term sums to 1 (Hoffman and Schadt, 2016; Nakagawa and Schielzeth, 2012).
</p>


<h3>Value</h3>

<p>fraction of variance explained / ICC for each variable in the regression model
</p>


<h3>References</h3>


<p>Nakagawa S, Johnson PC, Schielzeth H (2017).
“The coefficient of determination R 2 and intra-class correlation coefficient from generalized linear mixed-effects models revisited and expanded.”
<em>Journal of the Royal Society Interface</em>, <b>14</b>(134), 20170213. <br><br></p>
<p>Nakagawa S, Schielzeth H (2013).
“A general and simple method for obtaining R2 from generalized linear mixed-effects models.”
<em>Methods in ecology and evolution</em>, <b>4</b>(2), 133–142.<br><br></p>
<p>McKelvey RD, Zavoina W (1975).
“A statistical model for the analysis of ordinal level dependent variables.”
<em>Journal of mathematical sociology</em>, <b>4</b>(1), 103–120.<br><br></p>
<p>DeMaris A (2002).
“Explained variance in logistic regression: A Monte Carlo study of proposed measures.”
<em>Sociological Methods &amp; Research</em>, <b>31</b>(1), 27–74.<br><br></p>
<p>Hoffman GE, Schadt EE (2016).
“variancePartition: interpreting drivers of variation in complex gene expression studies.”
<em>BMC bioinformatics</em>, <b>17</b>(1), 1–13.<br><br></p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">library(lme4)
data(varPartData)

# Linear mixed model
fit &lt;- lmer(geneExpr[1, ] ~ (1 | Tissue) + Age, info)
calcVarPart(fit)

# Linear model
# Note that the two models produce slightly different results
# This is expected: they are different statistical estimates
# of the same underlying value
fit &lt;- lm(geneExpr[1, ] ~ Tissue + Age, info)
calcVarPart(fit)

</code><code class="language-r">library<span class="token punctuation">(</span>lme4<span class="token punctuation">)</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Linear mixed model</span>
fit <span class="token operator">&lt;-</span> lmer<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">,</span> <span class="token punctuation">]</span> <span class="token operator">~</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span> <span class="token operator">+</span> Age<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># Linear model</span>
<span class="token comment"># Note that the two models produce slightly different results</span>
<span class="token comment"># This is expected: they are different statistical estimates</span>
<span class="token comment"># of the same underlying value</span>
fit <span class="token operator">&lt;-</span> lm<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">,</span> <span class="token punctuation">]</span> <span class="token operator">~</span> Tissue <span class="token operator">+</span> Age<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
calcVarPart<span class="token punctuation">(</span>fit<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="canCorPairs"><div class="page-main">
<a href="#canCorPairs" class="help-page-title"><h2>canCorPairs</h2></a>

<h3>Description</h3>

<p>Assess correlation between all pairs of variables in a formula
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">canCorPairs(formula, data, showWarnings = TRUE)
</code><code class="language-r">canCorPairs<span class="token punctuation">(</span>formula<span class="token punctuation">,</span> data<span class="token punctuation">,</span> showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>standard additive linear model formula (doesn't support random effects currently, so just change the syntax)</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p>data.frame with the data for the variables in the formula</p>
</td>
</tr>
<tr>
<td><code id="showWarnings">showWarnings</code></td>
<td>
<p>default to true</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Canonical Correlation Analysis (CCA) is similar to correlation between two vectors, except that CCA can accommodate matricies as well.  For a pair of variables, canCorPairs assesses the degree to which they co-vary and contain the same information.  Variables in the formula can be a continuous variable or a discrete variable expanded to a matrix (which is done in the backend of a regression model).  For a pair of variables, canCorPairs uses CCA to compute the correlation between these variables and returns the pairwise correlation matrix.
</p>
<p>Statistically, let rho be the array of correlation values returned by the standard R function cancor to compute CCA.  <code>canCorPairs()</code> returns <code>sqrt(mean(rho^2))</code>, which is the fraction of the maximum possible correlation.  When comparing a two vectors, or a vector and a matrix, this gives the save value as the absolute correlation.  When comparing two sets of categorical variables (i.e. expanded to two matricies), this is equivalent to Cramer's V statistic.
</p>
<p>Note that CCA returns correlation values between 0 and 1.
</p>


<h3>Value</h3>

<p>Matrix of correlation values between all pairs of variables.
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

# load simulated data:
data(varPartData)

# specify formula
form &lt;- ~ Individual + Tissue + Batch + Age + Height

# Compute Canonical Correlation Analysis (CCA)
# between all pairs of variables
# returns absolute correlation value
C &lt;- canCorPairs(form, info)

# Plot correlation matrix
plotCorrMatrix(C)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># specify formula</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Individual <span class="token operator">+</span> Tissue <span class="token operator">+</span> Batch <span class="token operator">+</span> Age <span class="token operator">+</span> Height

<span class="token comment"># Compute Canonical Correlation Analysis (CCA)</span>
<span class="token comment"># between all pairs of variables</span>
<span class="token comment"># returns absolute correlation value</span>
C <span class="token operator">&lt;-</span> canCorPairs<span class="token punctuation">(</span>form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Plot correlation matrix</span>
plotCorrMatrix<span class="token punctuation">(</span>C<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="classifyTestsF"><div class="page-main">
<a href="#classifyTestsF" class="help-page-title"><h2>Multiple Testing Genewise Across Contrasts</h2></a>

<h3>Description</h3>

<p>For each gene, classify a series of related t-statistics as up, down or not significant.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">classifyTestsF(object, ...)
</code><code class="language-r">classifyTestsF<span class="token punctuation">(</span>object<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>numeric matrix of t-statistics or an 'MArrayLM2' object from which the t-statistics may be extracted.</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>additional arguments</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Works like limma::classifyTestsF, except object can have a list of covariance matrices object$cov.coefficients.list, instead of just one in object$cov.coefficients
</p>


<h3>See Also</h3>

<p><code>limma::classifyTestsF</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="classifyTestsF-MArrayLM2-method"><div class="page-main">
<a href="#classifyTestsF-MArrayLM2-method" class="help-page-title"><h2>Multiple Testing Genewise Across Contrasts</h2></a>

<h3>Description</h3>

<p>For each gene, classify a series of related t-statistics as up, down or not significant.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM2'
classifyTestsF(
  object,
  cor.matrix = NULL,
  df = Inf,
  p.value = 0.01,
  fstat.only = FALSE
)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM2'</span>
classifyTestsF<span class="token punctuation">(</span>
  object<span class="token punctuation">,</span>
  cor.matrix <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  df <span class="token operator">=</span> <span class="token number">Inf</span><span class="token punctuation">,</span>
  p.value <span class="token operator">=</span> <span class="token number">0.01</span><span class="token punctuation">,</span>
  fstat.only <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>numeric matrix of t-statistics or an 'MArrayLM2' object from which the t-statistics may be extracted.</p>
</td>
</tr>
<tr>
<td><code id="cor.matrix">cor.matrix</code></td>
<td>
<p>covariance matrix of each row of t-statistics.  Defaults to the identity matrix.</p>
</td>
</tr>
<tr>
<td><code id="df">df</code></td>
<td>
<p>numeric vector giving the degrees of freedom for the t-statistics.  May have length 1 or length equal to the number of rows of tstat.</p>
</td>
</tr>
<tr>
<td><code id="p.value">p.value</code></td>
<td>
<p>numeric value between 0 and 1 giving the desired size of the test</p>
</td>
</tr>
<tr>
<td><code id="fstat.only">fstat.only</code></td>
<td>
<p>logical, if 'TRUE' then return the overall F-statistic as for 'FStat' instead of classifying the test results</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Works like limma::classifyTestsF, except object can have a list of covariance matrices object$cov.coefficients.list, instead of just one in object$cov.coefficients
</p>


<h3>See Also</h3>

<p><code>limma::classifyTestsF</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="colinearityScore"><div class="page-main">
<a href="#colinearityScore" class="help-page-title"><h2>Collinearity score</h2></a>

<h3>Description</h3>

<p>Collinearity score for a regression model indicating if variables are too highly correlated to give meaningful results
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">colinearityScore(fit)
</code><code class="language-r">colinearityScore<span class="token punctuation">(</span>fit<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation"><tr>
<td><code id="fit">fit</code></td>
<td>
<p>regression model fit from lm() or lmer()</p>
</td>
</tr></table>
<h3>Value</h3>

<p>Returns the collinearity score between 0 and 1, where a score &gt; 0.999 means the degree of collinearity is too high.  This function reports the correlation matrix between coefficient estimates for fixed effects.  The collinearity score is the maximum absolute correlation value of this matrix. Note that the values are the correlation between the parameter estimates, and not between the variables themselves.
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

# load simulated data:
data(varPartData)
#
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

res &lt;- fitVarPartModel(geneExpr[1:10, ], form, info)

# evaluate the collinearity score on the first model fit
# this reports the correlation matrix between coefficients estimates
# for fixed effects
# the collinearity score is the maximum absolute correlation value
# If the collinearity score &gt; .999 then the variance partition
# estimates may be problematic
# In that case, a least one variable should be omitted
colinearityScore(res[[1]])

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>
<span class="token comment">#</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

res <span class="token operator">&lt;-</span> fitVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># evaluate the collinearity score on the first model fit</span>
<span class="token comment"># this reports the correlation matrix between coefficients estimates</span>
<span class="token comment"># for fixed effects</span>
<span class="token comment"># the collinearity score is the maximum absolute correlation value</span>
<span class="token comment"># If the collinearity score &gt; .999 then the variance partition</span>
<span class="token comment"># estimates may be problematic</span>
<span class="token comment"># In that case, a least one variable should be omitted</span>
colinearityScore<span class="token punctuation">(</span>res<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="deviation-method"><div class="page-main">
<a href="#deviation-method" class="help-page-title"><h2>Deviation from expectation for each observation</h2></a>

<h3>Description</h3>

<p>Given a model fit for each features, residuals are computed and transformed based on an absolute value or squaring transform.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">deviation(fit, method = c("AD", "SQ"), scale = c("leverage", "none"))

## S4 method for signature 'MArrayLM'
deviation(fit, method = c("AD", "SQ"), scale = c("leverage", "none"))
</code><code class="language-r">deviation<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"AD"</span><span class="token punctuation">,</span> <span class="token string">"SQ"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> scale <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"leverage"</span><span class="token punctuation">,</span> <span class="token string">"none"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM'</span>
deviation<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"AD"</span><span class="token punctuation">,</span> <span class="token string">"SQ"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> scale <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"leverage"</span><span class="token punctuation">,</span> <span class="token string">"none"</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit from <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="method">method</code></td>
<td>
<p>transform the residuals using absolute deviation ("AD") or squared deviation ("SQ").</p>
</td>
</tr>
<tr>
<td><code id="scale">scale</code></td>
<td>
<p>scale each observation by "leverage", or no scaling ("none")</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>matrix of deviations from expection for each observation
</p>


<h3>See Also</h3>

<p><code>diffVar()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)
library(edgeR)
data(varPartDEdata)

# filter genes by number of counts
isexpr &lt;- rowSums(cpm(countMatrix) &gt; 0.1) &gt;= 5

# Standard usage of limma/voom
dge &lt;- DGEList(countMatrix[isexpr, ])
dge &lt;- calcNormFactors(dge)

# make this vignette faster by analyzing a subset of genes
dge &lt;- dge[1:1000, ]

# regression formula
form &lt;- ~Disease

# estimate precision weights
vobj &lt;- voomWithDreamWeights(dge, form, metadata)

# fit dream model
fit &lt;- dream(vobj, form, metadata)
fit &lt;- eBayes(fit)

# Compute deviation from expection for each observation
# using model residuals
z &lt;- deviation(fit)
z[1:4, 1:4]

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>
library<span class="token punctuation">(</span>edgeR<span class="token punctuation">)</span>
data<span class="token punctuation">(</span>varPartDEdata<span class="token punctuation">)</span>

<span class="token comment"># filter genes by number of counts</span>
isexpr <span class="token operator">&lt;-</span> rowSums<span class="token punctuation">(</span>cpm<span class="token punctuation">(</span>countMatrix<span class="token punctuation">)</span> <span class="token operator">&gt;</span> <span class="token number">0.1</span><span class="token punctuation">)</span> <span class="token operator">&gt;=</span> <span class="token number">5</span>

<span class="token comment"># Standard usage of limma/voom</span>
dge <span class="token operator">&lt;-</span> DGEList<span class="token punctuation">(</span>countMatrix<span class="token punctuation">[</span>isexpr<span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">)</span>
dge <span class="token operator">&lt;-</span> calcNormFactors<span class="token punctuation">(</span>dge<span class="token punctuation">)</span>

<span class="token comment"># make this vignette faster by analyzing a subset of genes</span>
dge <span class="token operator">&lt;-</span> dge<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">1000</span><span class="token punctuation">,</span> <span class="token punctuation">]</span>

<span class="token comment"># regression formula</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span>Disease

<span class="token comment"># estimate precision weights</span>
vobj <span class="token operator">&lt;-</span> voomWithDreamWeights<span class="token punctuation">(</span>dge<span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>

<span class="token comment"># fit dream model</span>
fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>vobj<span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># Compute deviation from expection for each observation</span>
<span class="token comment"># using model residuals</span>
z <span class="token operator">&lt;-</span> deviation<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>
z<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">4</span><span class="token punctuation">,</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">4</span><span class="token punctuation">]</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="diffVar-method"><div class="page-main">
<a href="#diffVar-method" class="help-page-title"><h2>Test differential variance</h2></a>

<h3>Description</h3>

<p>Test the association between a covariate of interest and the response's deviation from expectation.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">diffVar(
  fit,
  method = c("AD", "SQ"),
  scale = c("leverage", "none"),
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'MArrayLM'
diffVar(
  fit,
  method = c("AD", "SQ"),
  scale = c("leverage", "none"),
  BPPARAM = SerialParam(),
  ...
)
</code><code class="language-r">diffVar<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"AD"</span><span class="token punctuation">,</span> <span class="token string">"SQ"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  scale <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"leverage"</span><span class="token punctuation">,</span> <span class="token string">"none"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM'</span>
diffVar<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"AD"</span><span class="token punctuation">,</span> <span class="token string">"SQ"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  scale <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"leverage"</span><span class="token punctuation">,</span> <span class="token string">"none"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit from <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="method">method</code></td>
<td>
<p>transform the residuals using absolute deviation ("AD") or squared deviation ("SQ").</p>
</td>
</tr>
<tr>
<td><code id="scale">scale</code></td>
<td>
<p>scale each observation by "leverage", or no scaling ("none")</p>
</td>
</tr>
<tr>
<td><code id="BPPARAM">BPPARAM</code></td>
<td>
<p>parameters for parallel evaluation</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other parameters passed to <code>dream()</code></p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>This method performs a test of differential variance between two subsets of the data, in a way that generalizes to multiple categories, continuous variables and metrics of spread beyond variance.  For the two category test, this method is simular to Levene's test.  This model was adapted from Phipson, et al (2014), extended to linear mixed models, and adapted to be compatible with <code>dream()</code>.
</p>
<p>This method is composed of multiple steps where 1) a typical linear (mixed) model is fit with <code>dream()</code>, 2) residuals are computed and transformed based on an absolute value or squaring transform, 3) a second regression is performed with <code>dream()</code> to test if a variable is associated with increased deviation from expectation.  Both regression take advantage of the <code>dream()</code> linear (mixed) modelling framework followed by empirical Bayes shrinkage that extends the <code>limma::voom()</code> framework.
</p>
<p>Note that <code>diffVar()</code> takes the results of the first regression as a parameter to use as a starting point.
</p>


<h3>Value</h3>

<p><code>MArrayLM</code> object storing differential results to be passed to <code>topTable()</code>
</p>


<h3>References</h3>


<p>Phipson B, Oshlack A (2014).
“DiffVar: a new method for detecting differential variability with application to methylation in cancer and aging.”
<em>Genome biology</em>, <b>15</b>(9), 1–16.

</p>


<h3>See Also</h3>

<p><code>missMethyl::diffVar()</code>, <code>car::leveneTest()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)
library(edgeR)
data(varPartDEdata)

# filter genes by number of counts
isexpr &lt;- rowSums(cpm(countMatrix) &gt; 0.1) &gt;= 5

# Standard usage of limma/voom
dge &lt;- DGEList(countMatrix[isexpr, ])
dge &lt;- calcNormFactors(dge)

# make this vignette faster by analyzing a subset of genes
dge &lt;- dge[1:1000, ]

# regression formula
form &lt;- ~Disease

# estimate precision weights
vobj &lt;- voomWithDreamWeights(dge, form, metadata)

# fit dream model
fit &lt;- dream(vobj, form, metadata)
fit &lt;- eBayes(fit)

# fit differential variance model
res &lt;- diffVar(fit)

# extract results for differential variance based on Disease
topTable(res, coef = "Disease1", number = 3)

# Box plot of top hit
# Since ASCL3 has a negative logFC,
# the deviation from expectation is *smaller* in
# Disease==1 compared to baseline.
gene &lt;- "ENST00000325884.1 gene=ASCL3"
boxplot(vobj$E[gene, ] ~ metadata$Disease, main = gene)

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>
library<span class="token punctuation">(</span>edgeR<span class="token punctuation">)</span>
data<span class="token punctuation">(</span>varPartDEdata<span class="token punctuation">)</span>

<span class="token comment"># filter genes by number of counts</span>
isexpr <span class="token operator">&lt;-</span> rowSums<span class="token punctuation">(</span>cpm<span class="token punctuation">(</span>countMatrix<span class="token punctuation">)</span> <span class="token operator">&gt;</span> <span class="token number">0.1</span><span class="token punctuation">)</span> <span class="token operator">&gt;=</span> <span class="token number">5</span>

<span class="token comment"># Standard usage of limma/voom</span>
dge <span class="token operator">&lt;-</span> DGEList<span class="token punctuation">(</span>countMatrix<span class="token punctuation">[</span>isexpr<span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">)</span>
dge <span class="token operator">&lt;-</span> calcNormFactors<span class="token punctuation">(</span>dge<span class="token punctuation">)</span>

<span class="token comment"># make this vignette faster by analyzing a subset of genes</span>
dge <span class="token operator">&lt;-</span> dge<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">1000</span><span class="token punctuation">,</span> <span class="token punctuation">]</span>

<span class="token comment"># regression formula</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span>Disease

<span class="token comment"># estimate precision weights</span>
vobj <span class="token operator">&lt;-</span> voomWithDreamWeights<span class="token punctuation">(</span>dge<span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>

<span class="token comment"># fit dream model</span>
fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>vobj<span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># fit differential variance model</span>
res <span class="token operator">&lt;-</span> diffVar<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># extract results for differential variance based on Disease</span>
topTable<span class="token punctuation">(</span>res<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Disease1"</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span>

<span class="token comment"># Box plot of top hit</span>
<span class="token comment"># Since ASCL3 has a negative logFC,</span>
<span class="token comment"># the deviation from expectation is *smaller* in</span>
<span class="token comment"># Disease==1 compared to baseline.</span>
gene <span class="token operator">&lt;-</span> <span class="token string">"ENST00000325884.1 gene=ASCL3"</span>
boxplot<span class="token punctuation">(</span>vobj<span class="token operator">$</span>E<span class="token punctuation">[</span>gene<span class="token punctuation">,</span> <span class="token punctuation">]</span> <span class="token operator">~</span> metadata<span class="token operator">$</span>Disease<span class="token punctuation">,</span> main <span class="token operator">=</span> gene<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="dream-method"><div class="page-main">
<a href="#dream-method" class="help-page-title"><h2>Differential expression with linear mixed model</h2></a>

<h3>Description</h3>

<p>Fit linear mixed model for differential expression and preform hypothesis test on fixed effects as specified in the contrast matrix <code>L</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">dream(
  exprObj,
  formula,
  data,
  L,
  ddf = c("adaptive", "Satterthwaite", "Kenward-Roger"),
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  REML = TRUE,
  BPPARAM = SerialParam(),
  ...
)
</code><code class="language-r">dream<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  L<span class="token punctuation">,</span>
  ddf <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"adaptive"</span><span class="token punctuation">,</span> <span class="token string">"Satterthwaite"</span><span class="token punctuation">,</span> <span class="token string">"Kenward-Roger"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="exprObj">exprObj</code></td>
<td>
<p>matrix of expression data (g genes x n samples), or <code>ExpressionSet</code>, or <code>EList</code> returned by voom() from the limma package</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work, and <code>lmFit()</code> followed by <code>contrasts.fit()</code> are run.</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p>data.frame with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="L">L</code></td>
<td>
<p>contrast matrix specifying a linear combination of fixed effects to test</p>
</td>
</tr>
<tr>
<td><code id="ddf">ddf</code></td>
<td>
<p>Specifiy "Satterthwaite" or "Kenward-Roger" method to estimate effective degress of freedom for hypothesis testing in the linear mixed model.  Note that Kenward-Roger is more accurate, but is *much* slower.  Satterthwaite is a good enough approximation for most datasets. "adaptive" (Default) uses KR for &lt;= 20 samples.</p>
</td>
</tr>
<tr>
<td><code id="useWeights">useWeights</code></td>
<td>
<p>if TRUE, analysis uses heteroskedastic error estimates from <code>voom()</code>.  Value is ignored unless exprObj is an <code>EList()</code> from <code>voom()</code> or <code>weightsMatrix</code> is specified</p>
</td>
</tr>
<tr>
<td><code id="control">control</code></td>
<td>
<p>control settings for <code>lmer()</code></p>
</td>
</tr>
<tr>
<td><code id="hideErrorsInBackend">hideErrorsInBackend</code></td>
<td>
<p>default FALSE.  If TRUE, hide errors in <code>attr(.,"errors")</code> and <code>attr(.,"error.initial")</code></p>
</td>
</tr>
<tr>
<td><code id="REML">REML</code></td>
<td>
<p>use restricted maximum likelihood to fit linear mixed model. default is TRUE.  See Details.</p>
</td>
</tr>
<tr>
<td><code id="BPPARAM">BPPARAM</code></td>
<td>
<p>parameters for parallel evaluation</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>Additional arguments for <code>lmer()</code> or <code>lm()</code></p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>A linear (mixed) model is fit for each gene in exprObj, using formula to specify variables in the regression (Hoffman and Roussos, 2021).  If categorical variables are modeled as random effects (as is recommended), then a linear mixed model us used.  For example if formula is <code>~ a + b + (1|c)</code>, then the model is
</p>
<p><code>fit &lt;- lmer( exprObj[j,] ~ a + b + (1|c), data=data)</code>
</p>
<p><code>useWeights=TRUE</code> causes <code>weightsMatrix[j,]</code> to be included as weights in the regression model.
</p>
<p>Note: Fitting the model for 20,000 genes can be computationally intensive.  To accelerate computation, models can be fit in parallel using <code>BiocParallel</code> to run code in parallel.  Parallel processing must be enabled before calling this function.  See below.
</p>
<p>The regression model is fit for each gene separately. Samples with missing values in either gene expression or metadata are omitted by the underlying call to lmer.
</p>
<p>Hypothesis tests and degrees of freedom are producted by <code>lmerTest</code> and <code>pbkrtest</code> pacakges
</p>
<p>While <code>REML=TRUE</code> is required by <code>lmerTest</code> when ddf='Kenward-Roger', ddf='Satterthwaite' can be used with <code>REML</code> as <code>TRUE</code> or <code>FALSE</code>.  Since the Kenward-Roger method gave the best power with an accurate control of false positive rate in our simulations, and since the Satterthwaite method with REML=TRUE gives p-values that are slightly closer to the Kenward-Roger p-values, <code>REML=TRUE</code> is the default.  See Vignette "3) Theory and practice of random effects and REML"
</p>


<h3>Value</h3>

<p>MArrayLM2 object (just like MArrayLM from limma), and the directly estimated p-value (without eBayes)
</p>


<h3>References</h3>


<p>Hoffman GE, Roussos P (2021).
“dream: Powerful differential expression analysis for repeated measures designs.”
<em>Bioinformatics</em>, <b>37</b>(2), 192–201.

</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)

# load simulated data:
# geneExpr: matrix of *normalized* gene expression values
# info: information/metadata about each sample
data(varPartData)

form &lt;- ~ Batch + (1 | Individual) + (1 | Tissue)

# Fit linear mixed model for each gene
# run on just 10 genes for time
# NOTE: dream() runs on *normalized* data
fit &lt;- dream(geneExpr[1:10, ], form, info)
fit &lt;- eBayes(fit)

# view top genes
topTable(fit, coef = "Batch2", number = 3)

# get contrast matrix testing if the coefficient for Batch3 is
# different from coefficient for Batch2
# Name this comparison as 'compare_3_2'
# The variable of interest must be a fixed effect
L &lt;- makeContrastsDream(form, info, contrasts = c(compare_3_2 = "Batch3 - Batch2"))

# plot contrasts
plotContrasts(L)

# Fit linear mixed model for each gene
# run on just 10 genes for time
fit2 &lt;- dream(geneExpr[1:10, ], form, info, L)
fit2 &lt;- eBayes(fit2)

# view top genes for this contrast
topTable(fit2, coef = "compare_3_2", number = 3)

# Parallel processing using multiple cores with reduced memory usage
param &lt;- SnowParam(4, "SOCK", progressbar = TRUE)
fit3 &lt;- dream(geneExpr[1:10, ], form, info, L, BPPARAM = param)
fit3 &lt;- eBayes(fit3)

# Fit fixed effect model for each gene
# Use lmFit in the backend
form &lt;- ~Batch
fit4 &lt;- dream(geneExpr[1:10, ], form, info, L)
fit4 &lt;- eBayes(fit4)

# view top genes
topTable(fit4, coef = "compare_3_2", number = 3)

# Compute residuals using dream
residuals(fit4)[1:4, 1:4]

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of *normalized* gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Fit linear mixed model for each gene</span>
<span class="token comment"># run on just 10 genes for time</span>
<span class="token comment"># NOTE: dream() runs on *normalized* data</span>
fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># view top genes</span>
topTable<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Batch2"</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span>

<span class="token comment"># get contrast matrix testing if the coefficient for Batch3 is</span>
<span class="token comment"># different from coefficient for Batch2</span>
<span class="token comment"># Name this comparison as 'compare_3_2'</span>
<span class="token comment"># The variable of interest must be a fixed effect</span>
L <span class="token operator">&lt;-</span> makeContrastsDream<span class="token punctuation">(</span>form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> contrasts <span class="token operator">=</span> c<span class="token punctuation">(</span>compare_3_2 <span class="token operator">=</span> <span class="token string">"Batch3 - Batch2"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># plot contrasts</span>
plotContrasts<span class="token punctuation">(</span>L<span class="token punctuation">)</span>

<span class="token comment"># Fit linear mixed model for each gene</span>
<span class="token comment"># run on just 10 genes for time</span>
fit2 <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> L<span class="token punctuation">)</span>
fit2 <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit2<span class="token punctuation">)</span>

<span class="token comment"># view top genes for this contrast</span>
topTable<span class="token punctuation">(</span>fit2<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"compare_3_2"</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span>

<span class="token comment"># Parallel processing using multiple cores with reduced memory usage</span>
param <span class="token operator">&lt;-</span> SnowParam<span class="token punctuation">(</span><span class="token number">4</span><span class="token punctuation">,</span> <span class="token string">"SOCK"</span><span class="token punctuation">,</span> progressbar <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
fit3 <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> L<span class="token punctuation">,</span> BPPARAM <span class="token operator">=</span> param<span class="token punctuation">)</span>
fit3 <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit3<span class="token punctuation">)</span>

<span class="token comment"># Fit fixed effect model for each gene</span>
<span class="token comment"># Use lmFit in the backend</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span>Batch
fit4 <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> L<span class="token punctuation">)</span>
fit4 <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit4<span class="token punctuation">)</span>

<span class="token comment"># view top genes</span>
topTable<span class="token punctuation">(</span>fit4<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"compare_3_2"</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span>

<span class="token comment"># Compute residuals using dream</span>
residuals<span class="token punctuation">(</span>fit4<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">4</span><span class="token punctuation">,</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">4</span><span class="token punctuation">]</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="dscchisq"><div class="page-main">
<a href="#dscchisq" class="help-page-title"><h2>Scaled chi-square</h2></a>

<h3>Description</h3>

<p>Scaled chi-square density using a gamma distribution
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">dscchisq(x, a, b)
</code><code class="language-r">dscchisq<span class="token punctuation">(</span>x<span class="token punctuation">,</span> a<span class="token punctuation">,</span> b<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="x">x</code></td>
<td>
<p>vector of quantiles.</p>
</td>
</tr>
<tr>
<td><code id="a">a</code></td>
<td>
<p>scale</p>
</td>
</tr>
<tr>
<td><code id="b">b</code></td>
<td>
<p>degrees of freedom</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="eBayes"><div class="page-main">
<a href="#eBayes" class="help-page-title"><h2>eBayes generic for for MArrayLM and MArrayLM2</h2></a>

<h3>Description</h3>

<p>eBayes for result of linear (mixed) model
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">eBayes(fit, ...)
</code><code class="language-r">eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>fit</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>all other args</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="eBayes-MArrayLM-method"><div class="page-main">
<a href="#eBayes-MArrayLM-method" class="help-page-title"><h2>eBayes for MArrayLM</h2></a>

<h3>Description</h3>

<p>eBayes for result of linear  model for with uses <code>limma::eBayes()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM'
eBayes(fit, ...)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM'</span>
eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>fit</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>all other args</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="eBayes-method"><div class="page-main">
<a href="#eBayes-method" class="help-page-title"><h2>eBayes for MArrayLM2</h2></a>

<h3>Description</h3>

<p>eBayes for result of linear mixed model for with <code>dream()</code> using residual degrees of freedom approximated with <code>rdf.merMod()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM2'
eBayes(
  fit,
  proportion = 0.01,
  stdev.coef.lim = c(0.1, 4),
  trend = FALSE,
  span = NULL,
  robust = FALSE,
  winsor.tail.p = c(0.05, 0.1),
  legacy = NULL
)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM2'</span>
eBayes<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  proportion <span class="token operator">=</span> <span class="token number">0.01</span><span class="token punctuation">,</span>
  stdev.coef.lim <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token number">0.1</span><span class="token punctuation">,</span> <span class="token number">4</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  trend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  span <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  robust <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  winsor.tail.p <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token number">0.05</span><span class="token punctuation">,</span> <span class="token number">0.1</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  legacy <span class="token operator">=</span> <span class="token keyword">NULL</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>fit</p>
</td>
</tr>
<tr>
<td><code id="proportion">proportion</code></td>
<td>
<p>proportion</p>
</td>
</tr>
<tr>
<td><code id="stdev.coef.lim">stdev.coef.lim</code></td>
<td>
<p>stdev.coef.lim</p>
</td>
</tr>
<tr>
<td><code id="trend">trend</code></td>
<td>
<p>trend</p>
</td>
</tr>
<tr>
<td><code id="span">span</code></td>
<td>
<p>span</p>
</td>
</tr>
<tr>
<td><code id="robust">robust</code></td>
<td>
<p>robust</p>
</td>
</tr>
<tr>
<td><code id="winsor.tail.p">winsor.tail.p</code></td>
<td>
<p>winsor.tail.p</p>
</td>
</tr>
<tr>
<td><code id="legacy">legacy</code></td>
<td>
<p>legacy</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>results of eBayes using approximated residual degrees of freedom
</p>


<h3>See Also</h3>

<p><code>dream()</code>, <code>rdf.merMod()</code>, <code>limma::eBayes()</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="ESS-method"><div class="page-main">
<a href="#ESS-method" class="help-page-title"><h2>Effective sample size</h2></a>

<h3>Description</h3>

<p>Compute effective sample size based on correlation structure in linear mixed model
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">ESS(fit, method = "full")

## S4 method for signature 'lmerMod'
ESS(fit, method = "full")
</code><code class="language-r">ESS<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> method <span class="token operator">=</span> <span class="token string">"full"</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'lmerMod'</span>
ESS<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> method <span class="token operator">=</span> <span class="token string">"full"</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit from lmer()</p>
</td>
</tr>
<tr>
<td><code id="method">method</code></td>
<td>
<p>"full" uses the full correlation structure of the model. The "approximate" method makes the simplifying assumption that the study has a mean of m samples in each of k groups, and computes m based on the study design.  When the study design is evenly balanced (i.e. the assumption is met), this gives the same results as the "full" method.</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Effective sample size calculations are based on:
</p>
<p>Liu, G., and Liang, K. Y. (1997). Sample size calculations for studies with correlated observations. Biometrics, 53(3), 937-47.
</p>
<p>"full" method: if </p>
<p style="text-align: center;"><code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>V</mi><mi>x</mi></msub><mo>=</mo><mi>v</mi><mi>a</mi><mi>r</mi><mo stretchy="false">(</mo><mi>Y</mi><mo separator="true">;</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">V_x = var(Y;x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.22222em;">V</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.2222em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">x</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.22222em;">Y</span><span class="mpunct">;</span><span class="mspace" style="margin-right:0.1667em;"></span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></code>
</p>
<p> is the variance-covariance matrix of Y, the response, based on the covariate x, then the effective sample size corresponding to this covariate is </p>
<p style="text-align: center;"><code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi mathvariant="normal">Σ</mi><mrow><mi>i</mi><mo separator="true">,</mo><mi>j</mi></mrow></msub><mo stretchy="false">(</mo><msubsup><mi>V</mi><mi>x</mi><mrow><mo>−</mo><mn>1</mn></mrow></msubsup><msub><mo stretchy="false">)</mo><mrow><mi>i</mi><mo separator="true">,</mo><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\Sigma_{i,j} (V_x^{-1})_{i,j}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.1002em;vertical-align:-0.2861em;"></span><span class="mord"><span class="mord">Σ</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3117em;"><span style="top:-2.55em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.05724em;">j</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.2861em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.22222em;">V</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-2.453em;margin-left:-0.2222em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">x</span></span></span><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">−</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.247em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3117em;"><span style="top:-2.55em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.05724em;">j</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.2861em;"><span></span></span></span></span></span></span></span></span></span></code>
</p>
<p>.  In R notation, this is: <code>sum(solve(V_x))</code>.  In practice, this can be evaluted as sum(w), where R 
</p>
<p>"approximate" method: Letting m be the mean number of samples per group, </p>
<p style="text-align: center;"><code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi></mrow><annotation encoding="application/x-tex">k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em;"></span><span class="mord mathnormal" style="margin-right:0.03148em;">k</span></span></span></span></code>
</p>
<p> be the number of groups, and </p>
<p style="text-align: center;"><code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ρ</mi></mrow><annotation encoding="application/x-tex">\rho</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.1944em;"></span><span class="mord mathnormal">ρ</span></span></span></span></code>
</p>
<p> be the intraclass correlation, the effective sample size is </p>
<p style="text-align: center;"><code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mi>k</mi><mi mathvariant="normal">/</mi><mo stretchy="false">(</mo><mn>1</mn><mo>+</mo><mi>ρ</mi><mo stretchy="false">(</mo><mi>m</mi><mo>−</mo><mn>1</mn><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">mk / (1+\rho(m-1))</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.03148em;">mk</span><span class="mord">/</span><span class="mopen">(</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">ρ</span><span class="mopen">(</span><span class="mord mathnormal">m</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">1</span><span class="mclose">))</span></span></span></span></code>
</p>

<p>Note that these values are equal when there are exactly m samples in each group.  If m is only an average then this an approximation.
</p>


<h3>Value</h3>

<p>effective sample size for each random effect in the model
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">library(lme4)
data(varPartData)

# Linear mixed model
fit &lt;- lmer(geneExpr[1, ] ~ (1 | Individual) + (1 | Tissue) + Age, info)

# Effective sample size
ESS(fit)

</code><code class="language-r">library<span class="token punctuation">(</span>lme4<span class="token punctuation">)</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Linear mixed model</span>
fit <span class="token operator">&lt;-</span> lmer<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">,</span> <span class="token punctuation">]</span> <span class="token operator">~</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span> <span class="token operator">+</span> Age<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Effective sample size</span>
ESS<span class="token punctuation">(</span>fit<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="extractVarPart"><div class="page-main">
<a href="#extractVarPart" class="help-page-title"><h2>Extract variance statistics</h2></a>

<h3>Description</h3>

<p>Extract variance statistics from list of models fit with <code>lm()</code> or <code>lmer()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">extractVarPart(modelList, ...)
</code><code class="language-r">extractVarPart<span class="token punctuation">(</span>modelList<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="modelList">modelList</code></td>
<td>
<p>list of <code>lmer()</code> model fits</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p><code>data.frame</code> of fraction of variance explained by each variable, after correcting for all others.
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Step 1: fit linear mixed model on gene expresson
# If categoritical variables are specified, a linear mixed model is used
# If all variables are modeled as continuous, a linear model is used
# each entry in results is a regression model fit on a single gene
# Step 2: extract variance fractions from each model fit
# for each gene, returns fraction of variation attributable to each variable
# Interpretation: the variance explained by each variable
# after correction for all other variables
varPart &lt;- fitExtractVarPartModel(geneExpr, form, info)

# violin plot of contribution of each variable to total variance
plotVarPart(sortCols(varPart))

# Advanced:
# Fit model and extract variance in two separate steps
# Step 1: fit model for each gene, store model fit for each gene in a list
results &lt;- fitVarPartModel(geneExpr, form, info)

# Step 2: extract variance fractions
varPart &lt;- extractVarPart(results)

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Step 1: fit linear mixed model on gene expresson</span>
<span class="token comment"># If categoritical variables are specified, a linear mixed model is used</span>
<span class="token comment"># If all variables are modeled as continuous, a linear model is used</span>
<span class="token comment"># each entry in results is a regression model fit on a single gene</span>
<span class="token comment"># Step 2: extract variance fractions from each model fit</span>
<span class="token comment"># for each gene, returns fraction of variation attributable to each variable</span>
<span class="token comment"># Interpretation: the variance explained by each variable</span>
<span class="token comment"># after correction for all other variables</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># violin plot of contribution of each variable to total variance</span>
plotVarPart<span class="token punctuation">(</span>sortCols<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># Advanced:</span>
<span class="token comment"># Fit model and extract variance in two separate steps</span>
<span class="token comment"># Step 1: fit model for each gene, store model fit for each gene in a list</span>
results <span class="token operator">&lt;-</span> fitVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Step 2: extract variance fractions</span>
varPart <span class="token operator">&lt;-</span> extractVarPart<span class="token punctuation">(</span>results<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="fitExtractVarPartModel-method"><div class="page-main">
<a href="#fitExtractVarPartModel-method" class="help-page-title"><h2>Fit linear (mixed) model, report variance fractions</h2></a>

<h3>Description</h3>

<p>Fit linear (mixed) model to estimate contribution of multiple sources of variation while simultaneously correcting for all other variables. Report fraction of variance attributable to each variable
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">fitExtractVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'matrix'
fitExtractVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'data.frame'
fitExtractVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'EList'
fitExtractVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'ExpressionSet'
fitExtractVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'sparseMatrix'
fitExtractVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)
</code><code class="language-r">fitExtractVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'matrix'</span>
fitExtractVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'data.frame'</span>
fitExtractVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'EList'</span>
fitExtractVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'ExpressionSet'</span>
fitExtractVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'sparseMatrix'</span>
fitExtractVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="exprObj">exprObj</code></td>
<td>
<p>matrix of expression data (g genes x n samples), or <code>ExpressionSet</code>, or <code>EList</code> returned by <code>voom()</code> from the <code>limma</code> package</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code></p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p><code>data.frame</code> with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="REML">REML</code></td>
<td>
<p>use restricted maximum likelihood to fit linear mixed model. default is FALSE.   See Details.</p>
</td>
</tr>
<tr>
<td><code id="useWeights">useWeights</code></td>
<td>
<p>if TRUE, analysis uses heteroskedastic error estimates from <code>voom()</code>.  Value is ignored unless exprObj is an <code>EList()</code> from <code>voom()</code> or <code>weightsMatrix</code> is specified</p>
</td>
</tr>
<tr>
<td><code id="control">control</code></td>
<td>
<p>control settings for <code>lmer()</code></p>
</td>
</tr>
<tr>
<td><code id="hideErrorsInBackend">hideErrorsInBackend</code></td>
<td>
<p>default FALSE.  If TRUE, hide errors in <code>attr(.,"errors")</code> and <code>attr(.,"error.initial")</code></p>
</td>
</tr>
<tr>
<td><code id="showWarnings">showWarnings</code></td>
<td>
<p>default TRUE. Indicate model failures</p>
</td>
</tr>
<tr>
<td><code id="BPPARAM">BPPARAM</code></td>
<td>
<p>parameters for parallel evaluation</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>Additional arguments for <code>lmer()</code> or <code>lm()</code></p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>A linear (mixed) model is fit for each gene in exprObj, using formula to specify variables in the regression.  If categorical variables are modeled as random effects (as is recommended), then a linear mixed model us used.  For example if formula is <code>~ a + b + (1|c)</code>, then the model is
</p>
<p>fit &lt;- lmer( exprObj[j,] ~ a + b + (1|c), data=data)
</p>
<p>If there are no random effects, so formula is ~ a + b + c, a 'standard' linear model is used:
</p>
<p><code>fit &lt;- lm( exprObj[j,] ~ a + b + c, data=data)</code>
</p>
<p>In both cases, <code>useWeights=TRUE</code> causes <code>weightsMatrix[j,]</code> to be included as weights in the regression model.
</p>
<p>Note: Fitting the model for 20,000 genes can be computationally intensive.  To accelerate computation, models can be fit in parallel using <code>BiocParallel</code> to run in parallel.  Parallel processing must be enabled before calling this function.  See below.
</p>
<p>The regression model is fit for each gene separately. Samples with missing values in either gene expression or metadata are omitted by the underlying call to <code>lm</code>/<code>lmer</code>.
</p>
<p><code>REML=FALSE</code> uses maximum likelihood to estimate variance fractions.  This approach produced unbiased estimates, while <code>REML=TRUE</code> can show substantial bias.  See Vignette "3) Theory and practice of random effects and REML"
</p>


<h3>Value</h3>

<p>list() of where each entry is a model fit produced by <code>lmer()</code> or <code>lm()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Step 1: fit linear mixed model on gene expression
# If categorical variables are specified, a linear mixed model is used
# If all variables are modeled as continuous, a linear model is used
# each entry in results is a regression model fit on a single gene
# Step 2: extract variance fractions from each model fit
# for each gene, returns fraction of variation attributable to each variable
# Interpretation: the variance explained by each variable
# after correction for all other variables
varPart &lt;- fitExtractVarPartModel(geneExpr, form, info)

# violin plot of contribution of each variable to total variance
plotVarPart(sortCols(varPart))

# Note: fitExtractVarPartModel also accepts ExpressionSet
data(sample.ExpressionSet, package = "Biobase")

# ExpressionSet example
form &lt;- ~ (1 | sex) + (1 | type) + score
info2 &lt;- Biobase::pData(sample.ExpressionSet)
varPart2 &lt;- fitExtractVarPartModel(sample.ExpressionSet, form, info2)


</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Step 1: fit linear mixed model on gene expression</span>
<span class="token comment"># If categorical variables are specified, a linear mixed model is used</span>
<span class="token comment"># If all variables are modeled as continuous, a linear model is used</span>
<span class="token comment"># each entry in results is a regression model fit on a single gene</span>
<span class="token comment"># Step 2: extract variance fractions from each model fit</span>
<span class="token comment"># for each gene, returns fraction of variation attributable to each variable</span>
<span class="token comment"># Interpretation: the variance explained by each variable</span>
<span class="token comment"># after correction for all other variables</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># violin plot of contribution of each variable to total variance</span>
plotVarPart<span class="token punctuation">(</span>sortCols<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># Note: fitExtractVarPartModel also accepts ExpressionSet</span>
data<span class="token punctuation">(</span>sample.ExpressionSet<span class="token punctuation">,</span> package <span class="token operator">=</span> <span class="token string">"Biobase"</span><span class="token punctuation">)</span>

<span class="token comment"># ExpressionSet example</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> sex<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> type<span class="token punctuation">)</span> <span class="token operator">+</span> score
info2 <span class="token operator">&lt;-</span> Biobase<span class="token operator">::</span>pData<span class="token punctuation">(</span>sample.ExpressionSet<span class="token punctuation">)</span>
varPart2 <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>sample.ExpressionSet<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info2<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="fitVarPartModel-method"><div class="page-main">
<a href="#fitVarPartModel-method" class="help-page-title"><h2>Fit linear (mixed) model</h2></a>

<h3>Description</h3>

<p>Fit linear (mixed) model to estimate contribution of multiple sources of variation while simultaneously correcting for all other variables.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">fitVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  fxn = identity,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'matrix'
fitVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  fxn = identity,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'data.frame'
fitVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  fxn = identity,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'EList'
fitVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  fxn = identity,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'ExpressionSet'
fitVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  fxn = identity,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'sparseMatrix'
fitVarPartModel(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  fxn = identity,
  control = vpcontrol,
  hideErrorsInBackend = FALSE,
  showWarnings = TRUE,
  BPPARAM = SerialParam(),
  ...
)
</code><code class="language-r">fitVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  fxn <span class="token operator">=</span> identity<span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'matrix'</span>
fitVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  fxn <span class="token operator">=</span> identity<span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'data.frame'</span>
fitVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  fxn <span class="token operator">=</span> identity<span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'EList'</span>
fitVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  fxn <span class="token operator">=</span> identity<span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'ExpressionSet'</span>
fitVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  fxn <span class="token operator">=</span> identity<span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'sparseMatrix'</span>
fitVarPartModel<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  fxn <span class="token operator">=</span> identity<span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  hideErrorsInBackend <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  showWarnings <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="exprObj">exprObj</code></td>
<td>
<p>matrix of expression data (g genes x n samples), or <code>ExpressionSet</code>, or <code>EList</code> returned by <code>voom()</code> from the <code>limma</code> package</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code></p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p><code>data.frame</code> with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="REML">REML</code></td>
<td>
<p>use restricted maximum likelihood to fit linear mixed model. default is FALSE.  See Details.</p>
</td>
</tr>
<tr>
<td><code id="useWeights">useWeights</code></td>
<td>
<p>if TRUE, analysis uses heteroskedastic error estimates from voom().  Value is ignored unless exprObj is an <code>EList()</code> from <code>voom()</code> or <code>weightsMatrix</code> is specified</p>
</td>
</tr>
<tr>
<td><code id="fxn">fxn</code></td>
<td>
<p>apply function to model fit for each gene.  Defaults to identify function so it returns the model fit itself</p>
</td>
</tr>
<tr>
<td><code id="control">control</code></td>
<td>
<p>control settings for <code>lmer()</code></p>
</td>
</tr>
<tr>
<td><code id="hideErrorsInBackend">hideErrorsInBackend</code></td>
<td>
<p>default FALSE.  If TRUE, hide errors in <code>attr(.,"errors")</code> and <code>attr(.,"error.initial")</code></p>
</td>
</tr>
<tr>
<td><code id="showWarnings">showWarnings</code></td>
<td>
<p>default TRUE. Indicate model failures</p>
</td>
</tr>
<tr>
<td><code id="BPPARAM">BPPARAM</code></td>
<td>
<p>parameters for parallel evaluation</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>Additional arguments for <code>lmer()</code> or <code>lm()</code></p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>A linear (mixed) model is fit for each gene in exprObj, using formula to specify variables in the regression.  If categorical variables are modeled as random effects (as is recommended), then a linear mixed model us used.  For example if formula is <code>~ a + b + (1|c)</code>, then the model is
</p>
<p><code>fit &lt;- lmer( exprObj[j,] ~ a + b + (1|c), data=data)</code>
</p>
<p>If there are no random effects, so formula is <code>~ a + b + c</code>, a 'standard' linear model is used:
</p>
<p><code>fit &lt;- lm( exprObj[j,] ~ a + b + c, data=data)</code>
</p>
<p>In both cases, <code>useWeights=TRUE</code> causes <code>weightsMatrix[j,]</code> to be included as weights in the regression model.
</p>
<p>Note: Fitting the model for 20,000 genes can be computationally intensive.  To accelerate computation, models can be fit in parallel using <code>BiocParallel</code> to run in parallel.  Parallel processing must be enabled before calling this function.  See below.
</p>
<p>The regression model is fit for each gene separately. Samples with missing values in either gene expression or metadata are omitted by the underlying call to lm/lmer.
</p>
<p>Since this function returns a list of each model fit, using this function is slower and uses more memory than <code>fitExtractVarPartModel()</code>.
</p>
<p><code>REML=FALSE</code> uses maximum likelihood to estimate variance fractions.  This approach produced unbiased estimates, while <code>REML=TRUE</code> can show substantial bias.  See Vignette "3) Theory and practice of random effects and REML"
</p>


<h3>Value</h3>

<p><code>list()</code> of where each entry is a model fit produced by <code>lmer()</code> or <code>lm()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Step 1: fit linear mixed model on gene expression
# If categorical variables are specified, a linear mixed model is used
# If all variables are modeled as continuous, a linear model is used
# each entry in results is a regression model fit on a single gene
# Step 2: extract variance fractions from each model fit
# for each gene, returns fraction of variation attributable to each variable
# Interpretation: the variance explained by each variable
# after correction for all other variables
varPart &lt;- fitExtractVarPartModel(geneExpr, form, info)

# violin plot of contribution of each variable to total variance
# also sort columns
plotVarPart(sortCols(varPart))

# Advanced:
# Fit model and extract variance in two separate steps
# Step 1: fit model for each gene, store model fit for each gene in a list
results &lt;- fitVarPartModel(geneExpr, form, info)

# Step 2: extract variance fractions
varPart &lt;- extractVarPart(results)

# Note: fitVarPartModel also accepts ExpressionSet
data(sample.ExpressionSet, package = "Biobase")

# ExpressionSet example
form &lt;- ~ (1 | sex) + (1 | type) + score
info2 &lt;- Biobase::pData(sample.ExpressionSet)
results2 &lt;- fitVarPartModel(sample.ExpressionSet, form, info2)


</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Step 1: fit linear mixed model on gene expression</span>
<span class="token comment"># If categorical variables are specified, a linear mixed model is used</span>
<span class="token comment"># If all variables are modeled as continuous, a linear model is used</span>
<span class="token comment"># each entry in results is a regression model fit on a single gene</span>
<span class="token comment"># Step 2: extract variance fractions from each model fit</span>
<span class="token comment"># for each gene, returns fraction of variation attributable to each variable</span>
<span class="token comment"># Interpretation: the variance explained by each variable</span>
<span class="token comment"># after correction for all other variables</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># violin plot of contribution of each variable to total variance</span>
<span class="token comment"># also sort columns</span>
plotVarPart<span class="token punctuation">(</span>sortCols<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># Advanced:</span>
<span class="token comment"># Fit model and extract variance in two separate steps</span>
<span class="token comment"># Step 1: fit model for each gene, store model fit for each gene in a list</span>
results <span class="token operator">&lt;-</span> fitVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Step 2: extract variance fractions</span>
varPart <span class="token operator">&lt;-</span> extractVarPart<span class="token punctuation">(</span>results<span class="token punctuation">)</span>

<span class="token comment"># Note: fitVarPartModel also accepts ExpressionSet</span>
data<span class="token punctuation">(</span>sample.ExpressionSet<span class="token punctuation">,</span> package <span class="token operator">=</span> <span class="token string">"Biobase"</span><span class="token punctuation">)</span>

<span class="token comment"># ExpressionSet example</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> sex<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> type<span class="token punctuation">)</span> <span class="token operator">+</span> score
info2 <span class="token operator">&lt;-</span> Biobase<span class="token operator">::</span>pData<span class="token punctuation">(</span>sample.ExpressionSet<span class="token punctuation">)</span>
results2 <span class="token operator">&lt;-</span> fitVarPartModel<span class="token punctuation">(</span>sample.ExpressionSet<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info2<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="get_prediction-method"><div class="page-main">
<a href="#get_prediction-method" class="help-page-title"><h2>Compute predicted value of formula for linear (mixed) model</h2></a>

<h3>Description</h3>

<p>Compute predicted value of formula for linear (mixed) model for with <code>lm</code> or <code>lmer</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">get_prediction(fit, formula)

## S4 method for signature 'lmerMod'
get_prediction(fit, formula)

## S4 method for signature 'lm'
get_prediction(fit, formula)
</code><code class="language-r">get_prediction<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> formula<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'lmerMod'</span>
get_prediction<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> formula<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'lm'</span>
get_prediction<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> formula<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit with <code>lm</code> or <code>lmer</code></p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>formula of fixed and random effects to predict</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Similar motivation as <code>lme4:::predict.merMod()</code>, but that function cannot use just a subset of the fixed effects: it either uses none or all.  Note that the intercept is included in the formula by default.  To exclude it from the prediction use <code>~ 0 + ...</code> syntax
</p>


<h3>Value</h3>

<p>Predicted values from formula using parameter estimates from fit linear (mixed) model
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
library(lme4)

# Linear model
fit &lt;- lm(Reaction ~ Days, sleepstudy)

# prediction of intercept
get_prediction(fit, ~1)

# prediction of Days without intercept
get_prediction(fit, ~ 0 + Days)

# Linear mixed model

# fit model
fm1 &lt;- lmer(Reaction ~ Days + (Days | Subject), sleepstudy)

# predict Days, but exclude intercept
get_prediction(fm1, ~ 0 + Days)

# predict Days and (Days | Subject) random effect, but exclude intercept
get_prediction(fm1, ~ 0 + Days + (Days | Subject))

</code><code class="language-r">library<span class="token punctuation">(</span>lme4<span class="token punctuation">)</span>

<span class="token comment"># Linear model</span>
fit <span class="token operator">&lt;-</span> lm<span class="token punctuation">(</span>Reaction <span class="token operator">~</span> Days<span class="token punctuation">,</span> sleepstudy<span class="token punctuation">)</span>

<span class="token comment"># prediction of intercept</span>
get_prediction<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token operator">~</span><span class="token number">1</span><span class="token punctuation">)</span>

<span class="token comment"># prediction of Days without intercept</span>
get_prediction<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token operator">~</span> <span class="token number">0</span> <span class="token operator">+</span> Days<span class="token punctuation">)</span>

<span class="token comment"># Linear mixed model</span>

<span class="token comment"># fit model</span>
fm1 <span class="token operator">&lt;-</span> lmer<span class="token punctuation">(</span>Reaction <span class="token operator">~</span> Days <span class="token operator">+</span> <span class="token punctuation">(</span>Days <span class="token operator">|</span> Subject<span class="token punctuation">)</span><span class="token punctuation">,</span> sleepstudy<span class="token punctuation">)</span>

<span class="token comment"># predict Days, but exclude intercept</span>
get_prediction<span class="token punctuation">(</span>fm1<span class="token punctuation">,</span> <span class="token operator">~</span> <span class="token number">0</span> <span class="token operator">+</span> Days<span class="token punctuation">)</span>

<span class="token comment"># predict Days and (Days | Subject) random effect, but exclude intercept</span>
get_prediction<span class="token punctuation">(</span>fm1<span class="token punctuation">,</span> <span class="token operator">~</span> <span class="token number">0</span> <span class="token operator">+</span> Days <span class="token operator">+</span> <span class="token punctuation">(</span>Days <span class="token operator">|</span> Subject<span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="getContrast-method"><div class="page-main">
<a href="#getContrast-method" class="help-page-title"><h2>Extract contrast matrix for linear mixed model</h2></a>

<h3>Description</h3>

<p>Extract contrast matrix, L, testing a single variable.  Contrasts involving more than one variable can be constructed by modifying L directly
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">getContrast(exprObj, formula, data, coefficient)
</code><code class="language-r">getContrast<span class="token punctuation">(</span>exprObj<span class="token punctuation">,</span> formula<span class="token punctuation">,</span> data<span class="token punctuation">,</span> coefficient<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="exprObj">exprObj</code></td>
<td>
<p>matrix of expression data (g genes x n samples), or <code>ExpressionSet</code>, or <code>EList</code> returned by <code>voom()</code> from the <code>limma</code> package</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p>data.frame with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="coefficient">coefficient</code></td>
<td>
<p>the coefficient to use in the hypothesis test</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>Contrast matrix testing one variable
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# get contrast matrix testing if the coefficient for Batch2 is zero
# The variable of interest must be a fixed effect
form &lt;- ~ Batch + (1 | Individual) + (1 | Tissue)
L &lt;- getContrast(geneExpr, form, info, "Batch3")

# get contrast matrix testing if Batch3 - Batch2 = 0
form &lt;- ~ Batch + (1 | Individual) + (1 | Tissue)
L &lt;- getContrast(geneExpr, form, info, c("Batch3", "Batch2"))

# To test against Batch1 use the formula:
# 	~ 0 + Batch + (1|Individual) + (1|Tissue)
# to estimate Batch1 directly instead of using it as the baseline

</code><code class="language-r"><span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># get contrast matrix testing if the coefficient for Batch2 is zero</span>
<span class="token comment"># The variable of interest must be a fixed effect</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>
L <span class="token operator">&lt;-</span> getContrast<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> <span class="token string">"Batch3"</span><span class="token punctuation">)</span>

<span class="token comment"># get contrast matrix testing if Batch3 - Batch2 = 0</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>
L <span class="token operator">&lt;-</span> getContrast<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> c<span class="token punctuation">(</span><span class="token string">"Batch3"</span><span class="token punctuation">,</span> <span class="token string">"Batch2"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># To test against Batch1 use the formula:</span>
<span class="token comment"># 	~ 0 + Batch + (1|Individual) + (1|Tissue)</span>
<span class="token comment"># to estimate Batch1 directly instead of using it as the baseline</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="getTreat-method"><div class="page-main">
<a href="#getTreat-method" class="help-page-title"><h2>Test if coefficient is different from a specified value</h2></a>

<h3>Description</h3>

<p>Test if coefficient is different from a specified value
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">getTreat(fit, lfc = log2(1.2), coef = 1, number = 10, sort.by = "p")

## S4 method for signature 'MArrayLM'
getTreat(fit, lfc = log2(1.2), coef = 1, number = 10, sort.by = "p")

## S4 method for signature 'MArrayLM2'
getTreat(fit, lfc = log2(1.2), coef = 1, number = 10, sort.by = "p")
</code><code class="language-r">getTreat<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> lfc <span class="token operator">=</span> log2<span class="token punctuation">(</span><span class="token number">1.2</span><span class="token punctuation">)</span><span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span> sort.by <span class="token operator">=</span> <span class="token string">"p"</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM'</span>
getTreat<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> lfc <span class="token operator">=</span> log2<span class="token punctuation">(</span><span class="token number">1.2</span><span class="token punctuation">)</span><span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span> sort.by <span class="token operator">=</span> <span class="token string">"p"</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM2'</span>
getTreat<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> lfc <span class="token operator">=</span> log2<span class="token punctuation">(</span><span class="token number">1.2</span><span class="token punctuation">)</span><span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span> sort.by <span class="token operator">=</span> <span class="token string">"p"</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>fit</p>
</td>
</tr>
<tr>
<td><code id="lfc">lfc</code></td>
<td>
<p>a minimum log2-fold-change below which changes not considered scientifically meaningful</p>
</td>
</tr>
<tr>
<td><code id="coef">coef</code></td>
<td>
<p>which coefficient to test</p>
</td>
</tr>
<tr>
<td><code id="number">number</code></td>
<td>
<p>number of genes to return</p>
</td>
</tr>
<tr>
<td><code id="sort.by">sort.by</code></td>
<td>
<p>column to sort by</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>results of getTreat
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
data(varPartData)

form &lt;- ~ Age + Batch + (1 | Individual) + (1 | Tissue)

fit &lt;- dream(geneExpr, form, info)
fit &lt;- eBayes(fit)

coef &lt;- "Age"

# Evaluate treat()/topTreat() in a way that works seamlessly for dream()
getTreat(fit, lfc = log2(1.03), coef, sort.by = "none", number = 3)

</code><code class="language-r">data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

coef <span class="token operator">&lt;-</span> <span class="token string">"Age"</span>

<span class="token comment"># Evaluate treat()/topTreat() in a way that works seamlessly for dream()</span>
getTreat<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> lfc <span class="token operator">=</span> log2<span class="token punctuation">(</span><span class="token number">1.03</span><span class="token punctuation">)</span><span class="token punctuation">,</span> coef<span class="token punctuation">,</span> sort.by <span class="token operator">=</span> <span class="token string">"none"</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="ggColorHue"><div class="page-main">
<a href="#ggColorHue" class="help-page-title"><h2>Default colors for ggplot</h2></a>

<h3>Description</h3>

<p>Return an array of n colors the same as the default used by ggplot2
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">ggColorHue(n)
</code><code class="language-r">ggColorHue<span class="token punctuation">(</span>n<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation"><tr>
<td><code id="n">n</code></td>
<td>
<p>number of colors</p>
</td>
</tr></table>
<h3>Value</h3>

<p>array of colors of length n
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">ggColorHue(4)
</code><code class="language-r">ggColorHue<span class="token punctuation">(</span><span class="token number">4</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="hatvalues-method"><div class="page-main">
<a href="#hatvalues-method" class="help-page-title"><h2>Compute hatvalues</h2></a>

<h3>Description</h3>

<p>Compute hatvalues from dream fit
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM'
hatvalues(model, vobj, ...)

## S4 method for signature 'MArrayLM2'
hatvalues(model, ...)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM'</span>
hatvalues<span class="token punctuation">(</span>model<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM2'</span>
hatvalues<span class="token punctuation">(</span>model<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="model">model</code></td>
<td>
<p>model fit from <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> returned by <code>voom()</code> or <code>voomWithDreamWeights()</code>.</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments, currently ignored</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="isRunableFormula"><div class="page-main">
<a href="#isRunableFormula" class="help-page-title"><h2>Test if formula is full rank on this dataset</h2></a>

<h3>Description</h3>

<p>Test if formula is full rank on this dataset
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">isRunableFormula(exprObj, formula, data)
</code><code class="language-r">isRunableFormula<span class="token punctuation">(</span>exprObj<span class="token punctuation">,</span> formula<span class="token punctuation">,</span> data<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="exprObj">exprObj</code></td>
<td>
<p>expression object</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>formula</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p>data</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="logLik.MArrayLM"><div class="page-main">
<a href="#logLik.MArrayLM" class="help-page-title"><h2>Log-likelihood from model fit</h2></a>

<h3>Description</h3>

<p>Log-likelihood from model fit
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S3 method for class 'MArrayLM'
logLik(object, vobj, ...)
</code><code class="language-r"><span class="token comment">## S3 method for class 'MArrayLM'</span>
logLik<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>result of <code>lmFit()</code> or <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> used to fit model</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>See <code>?stats::logLik</code></p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="logLik.MArrayLM2"><div class="page-main">
<a href="#logLik.MArrayLM2" class="help-page-title"><h2>Log-likelihood from model fit</h2></a>

<h3>Description</h3>

<p>Log-likelihood from model fit
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S3 method for class 'MArrayLM2'
logLik(object, ...)
</code><code class="language-r"><span class="token comment">## S3 method for class 'MArrayLM2'</span>
logLik<span class="token punctuation">(</span>object<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>result of <code>lmFit()</code> or <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>See <code>?stats::logLik</code></p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="makeContrastsDream"><div class="page-main">
<a href="#makeContrastsDream" class="help-page-title"><h2>Construct Matrix of Custom Contrasts</h2></a>

<h3>Description</h3>

<p>Construct the contrast matrix corresponding to specified contrasts of a set of parameters. Each specified set of contrast weights must sum to 1.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">makeContrastsDream(
  formula,
  data,
  ...,
  contrasts = NULL,
  suppressWarnings = FALSE,
  nullOnError = FALSE
)
</code><code class="language-r">makeContrastsDream<span class="token punctuation">(</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  <span class="token ellipsis">...</span><span class="token punctuation">,</span>
  contrasts <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  suppressWarnings <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  nullOnError <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p>data.frame with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>expressions, or character strings which can be parsed to expressions, specifying contrasts</p>
</td>
</tr>
<tr>
<td><code id="contrasts">contrasts</code></td>
<td>
<p>character vector specifying contrasts</p>
</td>
</tr>
<tr>
<td><code id="suppressWarnings">suppressWarnings</code></td>
<td>
<p>(default FALSE). suppress warnings for univariate contrasts</p>
</td>
</tr>
<tr>
<td><code id="nullOnError">nullOnError</code></td>
<td>
<p>(default FALSE). When a contrast entry is invalid, throw warning and return NULL for that contrast entry</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>This function expresses contrasts between a set of parameters as a numeric matrix. The parameters are usually the coefficients from a linear (mixed) model fit, so the matrix specifies which comparisons between the coefficients are to be extracted from the fit. The output from this function is usually used as input to <code>dream()</code>.
</p>
<p>This function creates a matrix storing the contrasts weights that are applied to each coefficient.
</p>
<p>Consider a variable <code>v</code> with levels <code>c('A', 'B', 'C')</code>.  A contrast comparing <code>A</code> and <code>B</code> is <code>'vA - vB'</code> and tests whether the difference between these levels is different than zero. Coded for the 3 levels this has weights <code>c(1, -1, 0)</code>.  In order to compare <code>A</code> to the other levels, the contrast is <code>'vA - (vB + vC)/2'</code> so that <code>A</code> is compared to the average of the other two levels. This is encoded as <code>c(1, -0.5, -0.5)</code>.  This type of proper matching in testing multiple levels is enforced by ensuring that the contrast weights sum to 1. Based on standard regression theory only weighted sums of the estimated coefficients are supported.
</p>
<p>This function is inspired by <code>limma::makeContrasts()</code> but is designed to be compatible with linear mixed models for <code>dream()</code>
</p>
<p>Names in ... and contrasts will be used as column names in the returned value.
</p>


<h3>Value</h3>

<p>matrix of linear contrasts between regression coefficients
</p>


<h3>See Also</h3>

<p><code>plotContrasts()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

form &lt;- ~ 0 + Batch + (1 | Individual) + (1 | Tissue)

# Define contrasts
# Note that for each contrass, the weights sum to 1
L &lt;- makeContrastsDream(form, info, contrasts = c(Batch1_vs_2 = "Batch1 - Batch2", Batch3_vs_4 = "Batch3 - Batch4", Batch1_vs_34 = "Batch1 - (Batch3 + Batch4)/2"))

# show contrasts matrix
L

# Plot to visualize contrasts matrix
plotContrasts(L)

# Fit linear mixed model for each gene
# run on just 10 genes for time
fit &lt;- dream(geneExpr[1:10, ], form, info, L = L)

# examine contrasts after fitting
head(coef(fit))

# show results from first contrast
topTable(fit, coef = "Batch1_vs_2")

# show results from second contrast
topTable(fit, coef = "Batch3_vs_4")

# show results from third contrast
topTable(fit, coef = "Batch1_vs_34")

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

form <span class="token operator">&lt;-</span> <span class="token operator">~</span> <span class="token number">0</span> <span class="token operator">+</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Define contrasts</span>
<span class="token comment"># Note that for each contrass, the weights sum to 1</span>
L <span class="token operator">&lt;-</span> makeContrastsDream<span class="token punctuation">(</span>form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> contrasts <span class="token operator">=</span> c<span class="token punctuation">(</span>Batch1_vs_2 <span class="token operator">=</span> <span class="token string">"Batch1 - Batch2"</span><span class="token punctuation">,</span> Batch3_vs_4 <span class="token operator">=</span> <span class="token string">"Batch3 - Batch4"</span><span class="token punctuation">,</span> Batch1_vs_34 <span class="token operator">=</span> <span class="token string">"Batch1 - (Batch3 + Batch4)/2"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># show contrasts matrix</span>
L

<span class="token comment"># Plot to visualize contrasts matrix</span>
plotContrasts<span class="token punctuation">(</span>L<span class="token punctuation">)</span>

<span class="token comment"># Fit linear mixed model for each gene</span>
<span class="token comment"># run on just 10 genes for time</span>
fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> L <span class="token operator">=</span> L<span class="token punctuation">)</span>

<span class="token comment"># examine contrasts after fitting</span>
head<span class="token punctuation">(</span>coef<span class="token punctuation">(</span>fit<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># show results from first contrast</span>
topTable<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Batch1_vs_2"</span><span class="token punctuation">)</span>

<span class="token comment"># show results from second contrast</span>
topTable<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Batch3_vs_4"</span><span class="token punctuation">)</span>

<span class="token comment"># show results from third contrast</span>
topTable<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Batch1_vs_34"</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="MArrayLM2-class"><div class="page-main">
<a href="#MArrayLM2-class" class="help-page-title"><h2>Class MArrayLM2</h2></a>

<h3>Description</h3>

<p>Class <code>MArrayLM2</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="mvTest-method"><div class="page-main">
<a href="#mvTest-method" class="help-page-title"><h2>Multivariate tests on results from <code>dream()</code>
</h2></a>

<h3>Description</h3>

<p>Evaluate multivariate tests on results from <code>dream()</code> using <code>vcov()</code> to compute the covariance between estimated regression coefficients across multiple responses.  A joint test to see if the coefficients are jointly different from zero is performed using meta-analysis methods that account for the covariance.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">mvTest(
  fit,
  vobj,
  features,
  coef,
  method = c("FE.empirical", "FE", "RE2C", "tstat", "hotelling", "sidak", "fisher"),
  shrink.cov = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'MArrayLM,EList,vector'
mvTest(
  fit,
  vobj,
  features,
  coef,
  method = c("FE.empirical", "FE", "RE2C", "tstat", "hotelling", "sidak", "fisher"),
  shrink.cov = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'MArrayLM,EList,missing'
mvTest(
  fit,
  vobj,
  features,
  coef,
  method = c("FE.empirical", "FE", "RE2C", "tstat", "hotelling", "sidak", "fisher"),
  shrink.cov = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'MArrayLM,EList,list'
mvTest(
  fit,
  vobj,
  features,
  coef,
  method = c("FE.empirical", "FE", "RE2C", "tstat", "hotelling", "sidak", "fisher"),
  shrink.cov = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'mvTest_input,ANY,ANY'
mvTest(
  fit,
  vobj,
  features,
  coef,
  method = c("FE.empirical", "FE", "RE2C", "tstat", "hotelling", "sidak", "fisher"),
  shrink.cov = TRUE,
  BPPARAM = SerialParam(),
  ...
)

## S4 method for signature 'MArrayLM,matrix,ANY'
mvTest(
  fit,
  vobj,
  features,
  coef,
  method = c("FE.empirical", "FE", "RE2C", "tstat", "hotelling", "sidak", "fisher"),
  shrink.cov = TRUE,
  BPPARAM = SerialParam(),
  ...
)
</code><code class="language-r">mvTest<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  vobj<span class="token punctuation">,</span>
  features<span class="token punctuation">,</span>
  coef<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"FE.empirical"</span><span class="token punctuation">,</span> <span class="token string">"FE"</span><span class="token punctuation">,</span> <span class="token string">"RE2C"</span><span class="token punctuation">,</span> <span class="token string">"tstat"</span><span class="token punctuation">,</span> <span class="token string">"hotelling"</span><span class="token punctuation">,</span> <span class="token string">"sidak"</span><span class="token punctuation">,</span> <span class="token string">"fisher"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  shrink.cov <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM,EList,vector'</span>
mvTest<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  vobj<span class="token punctuation">,</span>
  features<span class="token punctuation">,</span>
  coef<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"FE.empirical"</span><span class="token punctuation">,</span> <span class="token string">"FE"</span><span class="token punctuation">,</span> <span class="token string">"RE2C"</span><span class="token punctuation">,</span> <span class="token string">"tstat"</span><span class="token punctuation">,</span> <span class="token string">"hotelling"</span><span class="token punctuation">,</span> <span class="token string">"sidak"</span><span class="token punctuation">,</span> <span class="token string">"fisher"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  shrink.cov <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM,EList,missing'</span>
mvTest<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  vobj<span class="token punctuation">,</span>
  features<span class="token punctuation">,</span>
  coef<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"FE.empirical"</span><span class="token punctuation">,</span> <span class="token string">"FE"</span><span class="token punctuation">,</span> <span class="token string">"RE2C"</span><span class="token punctuation">,</span> <span class="token string">"tstat"</span><span class="token punctuation">,</span> <span class="token string">"hotelling"</span><span class="token punctuation">,</span> <span class="token string">"sidak"</span><span class="token punctuation">,</span> <span class="token string">"fisher"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  shrink.cov <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM,EList,list'</span>
mvTest<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  vobj<span class="token punctuation">,</span>
  features<span class="token punctuation">,</span>
  coef<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"FE.empirical"</span><span class="token punctuation">,</span> <span class="token string">"FE"</span><span class="token punctuation">,</span> <span class="token string">"RE2C"</span><span class="token punctuation">,</span> <span class="token string">"tstat"</span><span class="token punctuation">,</span> <span class="token string">"hotelling"</span><span class="token punctuation">,</span> <span class="token string">"sidak"</span><span class="token punctuation">,</span> <span class="token string">"fisher"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  shrink.cov <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'mvTest_input,ANY,ANY'</span>
mvTest<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  vobj<span class="token punctuation">,</span>
  features<span class="token punctuation">,</span>
  coef<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"FE.empirical"</span><span class="token punctuation">,</span> <span class="token string">"FE"</span><span class="token punctuation">,</span> <span class="token string">"RE2C"</span><span class="token punctuation">,</span> <span class="token string">"tstat"</span><span class="token punctuation">,</span> <span class="token string">"hotelling"</span><span class="token punctuation">,</span> <span class="token string">"sidak"</span><span class="token punctuation">,</span> <span class="token string">"fisher"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  shrink.cov <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM,matrix,ANY'</span>
mvTest<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  vobj<span class="token punctuation">,</span>
  features<span class="token punctuation">,</span>
  coef<span class="token punctuation">,</span>
  method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"FE.empirical"</span><span class="token punctuation">,</span> <span class="token string">"FE"</span><span class="token punctuation">,</span> <span class="token string">"RE2C"</span><span class="token punctuation">,</span> <span class="token string">"tstat"</span><span class="token punctuation">,</span> <span class="token string">"hotelling"</span><span class="token punctuation">,</span> <span class="token string">"sidak"</span><span class="token punctuation">,</span> <span class="token string">"fisher"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  shrink.cov <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p><code>MArrayLM</code> or <code>MArrayLM2</code> returned by <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p>matrix or <code>EList</code> object returned by <code>voom()</code></p>
</td>
</tr>
<tr>
<td><code id="features">features</code></td>
<td>
<p>a) indeces or names of features to perform multivariate test on, b) list of indeces or names.  If missing, perform joint test on all features.</p>
</td>
</tr>
<tr>
<td><code id="coef">coef</code></td>
<td>
<p>name of coefficient or contrast to be tested</p>
</td>
</tr>
<tr>
<td><code id="method">method</code></td>
<td>
<p>statistical method used to perform multivariate test.  See details.  <code>'FE'</code> is a fixed effect test that models the covariance between coefficients.  <code>'FE.empirical'</code> use compute empirical p-values by sampling from the null distribution and fitting with a gamma. <code>'RE2C'</code> is a random effect test of heterogeneity of the estimated coefficients that models the covariance between coefficients, and also incorporates a fixed effects test too. <code>'tstat'</code> combines the t-statistics and models the covariance between coefficients. <code>'hotelling'</code> performs the Hotelling T2 test. <code>'sidak'</code> returns the smallest p-value and accounting for the number of tests. <code>'fisher'</code> combines the p-value using Fisher's method assuming independent tests.</p>
</td>
</tr>
<tr>
<td><code id="shrink.cov">shrink.cov</code></td>
<td>
<p>shrink the covariance matrix between coefficients using the Schafer-Strimmer method</p>
</td>
</tr>
<tr>
<td><code id="BPPARAM">BPPARAM</code></td>
<td>
<p>parameters for parallel evaluation</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arugments</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>See package <code>remaCor</code> for details about the <code>remaCor::RE2C()</code> test, and see <code>remaCor::LS()</code> for details about the fixed effect test.  When only 1 feature is selected, the original p-value is returned and the test statistic is set to <code>NA</code>.
</p>
<p>For the <code>"RE2C"</code> test, the final test statistic is the sum of a test statistic for the mean effect (<code>stat.FE</code>) and heterogeneity across effects (<code>stat.het</code>).  <code>mvTest()</code> returns 0 if <code>stat.het</code> is negative in extremely rare cases.
</p>


<h3>Value</h3>

<p>Returns a <code>data.frame</code> with the statistics from each test, the <code>pvalue</code> from the test, <code>n_features</code>,  <code>method</code>, and <code>lambda</code> from the Schafer-Strimmer method to shrink the estimated covariance.  When <code>shrink.cov=FALSE</code>, <code>lambda = 0</code>.
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)
library(edgeR)
library(BiocParallel)

data(varPartDEdata)

# normalize RNA-seq counts
dge &lt;- DGEList(counts = countMatrix)
dge &lt;- calcNormFactors(dge)

# specify formula with random effect for Individual
form &lt;- ~ Disease + (1 | Individual)

# compute observation weights
vobj &lt;- voomWithDreamWeights(dge[1:20, ], form, metadata)

# fit dream model
fit &lt;- dream(vobj, form, metadata)
fit &lt;- eBayes(fit)

# Multivariate test of features 1 and 2
mvTest(fit, vobj, 1:2, coef = "Disease1")

# Test multiple sets of features
lst &lt;- list(a = 1:2, b = 3:4)
mvTest(fit, vobj, lst, coef = "Disease1", BPPARAM = SnowParam(2))
</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>
library<span class="token punctuation">(</span>edgeR<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartDEdata<span class="token punctuation">)</span>

<span class="token comment"># normalize RNA-seq counts</span>
dge <span class="token operator">&lt;-</span> DGEList<span class="token punctuation">(</span>counts <span class="token operator">=</span> countMatrix<span class="token punctuation">)</span>
dge <span class="token operator">&lt;-</span> calcNormFactors<span class="token punctuation">(</span>dge<span class="token punctuation">)</span>

<span class="token comment"># specify formula with random effect for Individual</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Disease <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span>

<span class="token comment"># compute observation weights</span>
vobj <span class="token operator">&lt;-</span> voomWithDreamWeights<span class="token punctuation">(</span>dge<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">20</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>

<span class="token comment"># fit dream model</span>
fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>vobj<span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># Multivariate test of features 1 and 2</span>
mvTest<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">2</span><span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Disease1"</span><span class="token punctuation">)</span>

<span class="token comment"># Test multiple sets of features</span>
lst <span class="token operator">&lt;-</span> list<span class="token punctuation">(</span>a <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">2</span><span class="token punctuation">,</span> b <span class="token operator">=</span> <span class="token number">3</span><span class="token operator">:</span><span class="token number">4</span><span class="token punctuation">)</span>
mvTest<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> lst<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Disease1"</span><span class="token punctuation">,</span> BPPARAM <span class="token operator">=</span> SnowParam<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="mvTest_input-class"><div class="page-main">
<a href="#mvTest_input-class" class="help-page-title"><h2>Class mvTest_input</h2></a>

<h3>Description</h3>

<p>Class <code>mvTest_input</code> work is with <code>iterRowsSplit()</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="plotCompareP-method"><div class="page-main">
<a href="#plotCompareP-method" class="help-page-title"><h2>Compare p-values from two analyses</h2></a>

<h3>Description</h3>

<p>Plot -log10 p-values from two analyses and color based on donor component from variancePartition analysis
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotCompareP(
  p1,
  p2,
  vpDonor,
  dupcorvalue,
  fraction = 0.2,
  xlabel = bquote(duplicateCorrelation ~ (-log[10] ~ p)),
  ylabel = bquote(dream ~ (-log[10] ~ p))
)
</code><code class="language-r">plotCompareP<span class="token punctuation">(</span>
  p1<span class="token punctuation">,</span>
  p2<span class="token punctuation">,</span>
  vpDonor<span class="token punctuation">,</span>
  dupcorvalue<span class="token punctuation">,</span>
  fraction <span class="token operator">=</span> <span class="token number">0.2</span><span class="token punctuation">,</span>
  xlabel <span class="token operator">=</span> bquote<span class="token punctuation">(</span>duplicateCorrelation <span class="token operator">~</span> <span class="token punctuation">(</span><span class="token operator">-</span>log<span class="token punctuation">[</span><span class="token number">10</span><span class="token punctuation">]</span> <span class="token operator">~</span> p<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  ylabel <span class="token operator">=</span> bquote<span class="token punctuation">(</span>dream <span class="token operator">~</span> <span class="token punctuation">(</span><span class="token operator">-</span>log<span class="token punctuation">[</span><span class="token number">10</span><span class="token punctuation">]</span> <span class="token operator">~</span> p<span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="p1">p1</code></td>
<td>
<p>p-value from first analysis</p>
</td>
</tr>
<tr>
<td><code id="p2">p2</code></td>
<td>
<p>p-value from second analysis</p>
</td>
</tr>
<tr>
<td><code id="vpDonor">vpDonor</code></td>
<td>
<p>donor component for each gene from variancePartition analysis</p>
</td>
</tr>
<tr>
<td><code id="dupcorvalue">dupcorvalue</code></td>
<td>
<p>scalar donor component from duplicateCorrelation</p>
</td>
</tr>
<tr>
<td><code id="fraction">fraction</code></td>
<td>
<p>fraction of highest/lowest values to use for best fit lines</p>
</td>
</tr>
<tr>
<td><code id="xlabel">xlabel</code></td>
<td>
<p>for x-axis</p>
</td>
</tr>
<tr>
<td><code id="ylabel">ylabel</code></td>
<td>
<p>label for y-axis</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>ggplot2 plot
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Perform very simple analysis for demonstration

# Analysis 1
form &lt;- ~Batch
fit &lt;- dream(geneExpr, form, info)
fit &lt;- eBayes(fit)
res &lt;- topTable(fit, number = Inf, coef = "Batch3")

# Analysis 2
form &lt;- ~ Batch + (1 | Tissue)
fit2 &lt;- dream(geneExpr, form, info)
res2 &lt;- topTable(fit2, number = Inf, coef = "Batch3")

# Compare p-values
plotCompareP(res$P.Value, res2$P.Value, runif(nrow(res)), .3)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Perform very simple analysis for demonstration</span>

<span class="token comment"># Analysis 1</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span>Batch
fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>
res <span class="token operator">&lt;-</span> topTable<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">Inf</span><span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Batch3"</span><span class="token punctuation">)</span>

<span class="token comment"># Analysis 2</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>
fit2 <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
res2 <span class="token operator">&lt;-</span> topTable<span class="token punctuation">(</span>fit2<span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">Inf</span><span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Batch3"</span><span class="token punctuation">)</span>

<span class="token comment"># Compare p-values</span>
plotCompareP<span class="token punctuation">(</span>res<span class="token operator">$</span>P.Value<span class="token punctuation">,</span> res2<span class="token operator">$</span>P.Value<span class="token punctuation">,</span> runif<span class="token punctuation">(</span>nrow<span class="token punctuation">(</span>res<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token number">.3</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotContrasts"><div class="page-main">
<a href="#plotContrasts" class="help-page-title"><h2>Plot representation of contrast matrix</h2></a>

<h3>Description</h3>

<p>Plot contrast matrix to clarify interpretation of hypothesis tests with linear contrasts
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotContrasts(L)
</code><code class="language-r">plotContrasts<span class="token punctuation">(</span>L<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation"><tr>
<td><code id="L">L</code></td>
<td>
<p>contrast matrix</p>
</td>
</tr></table>
<h3>Details</h3>

<p>This plot shows the contrasts weights that are applied to each coefficient.
</p>
<p>Consider a variable <code>v</code> with levels <code>c('A', 'B', 'C')</code>.  A contrast comparing <code>A</code> and <code>B</code> is <code>'vA - vB'</code> and tests whether the difference between these levels is different than zero. Coded for the 3 levels this has weights <code>c(1, -1, 0)</code>.  In order to compare <code>A</code> to the other levels, the contrast is <code>'vA - (vB + vC)/2'</code> so that <code>A</code> is compared to the average of the other two levels. This is encoded as <code>c(1, -0.5, -0.5)</code>.  This type of proper matching in testing multiple levels is enforced by ensuring that the contrast weights sum to 1. Based on standard regression theory only weighted sums of the estimated coefficients are supported.
</p>


<h3>Value</h3>

<p>ggplot2 object
</p>


<h3>See Also</h3>

<p><code>makeContrastsDream()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># load library
# library(variancePartition)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# 1) get contrast matrix testing if the coefficient for Batch2 is different from Batch3
form &lt;- ~ Batch + (1 | Individual) + (1 | Tissue)
L &lt;- makeContrastsDream(form, info, contrasts = c(Batch_3_vs_2 = "Batch3 - Batch2"))

# plot contrasts
plotContrasts(L)
</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># 1) get contrast matrix testing if the coefficient for Batch2 is different from Batch3</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Batch <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>
L <span class="token operator">&lt;-</span> makeContrastsDream<span class="token punctuation">(</span>form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> contrasts <span class="token operator">=</span> c<span class="token punctuation">(</span>Batch_3_vs_2 <span class="token operator">=</span> <span class="token string">"Batch3 - Batch2"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># plot contrasts</span>
plotContrasts<span class="token punctuation">(</span>L<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotCorrMatrix"><div class="page-main">
<a href="#plotCorrMatrix" class="help-page-title"><h2>plotCorrMatrix</h2></a>

<h3>Description</h3>

<p>Plot correlation matrix
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotCorrMatrix(
  C,
  dendrogram = "both",
  sort = TRUE,
  margins = c(13, 13),
  key.xlab = "correlation",
  ...
)
</code><code class="language-r">plotCorrMatrix<span class="token punctuation">(</span>
  C<span class="token punctuation">,</span>
  dendrogram <span class="token operator">=</span> <span class="token string">"both"</span><span class="token punctuation">,</span>
  sort <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  margins <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token number">13</span><span class="token punctuation">,</span> <span class="token number">13</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  key.xlab <span class="token operator">=</span> <span class="token string">"correlation"</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="C">C</code></td>
<td>
<p>correlation matrix: R or R^2 matrix</p>
</td>
</tr>
<tr>
<td><code id="dendrogram">dendrogram</code></td>
<td>
<p>character string indicating whether to draw 'both' or none'</p>
</td>
</tr>
<tr>
<td><code id="sort">sort</code></td>
<td>
<p>sort rows and columns based on clustering</p>
</td>
</tr>
<tr>
<td><code id="margins">margins</code></td>
<td>
<p>spacing of plot</p>
</td>
</tr>
<tr>
<td><code id="key.xlab">key.xlab</code></td>
<td>
<p>label of color gradient</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>additional arguments to heatmap.2</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Plots image of correlation matrix using customized call to heatmap.2
</p>


<h3>Value</h3>

<p>Image of correlation matrix
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# simulate simple matrix of 10 variables
mat &lt;- matrix(rnorm(1000), ncol = 10)

# compute correlation matrix
C &lt;- cor(mat)

# plot correlations
plotCorrMatrix(C)

# plot squared correlations
plotCorrMatrix(C^2, dendrogram = "none")

</code><code class="language-r"><span class="token comment"># simulate simple matrix of 10 variables</span>
mat <span class="token operator">&lt;-</span> matrix<span class="token punctuation">(</span>rnorm<span class="token punctuation">(</span><span class="token number">1000</span><span class="token punctuation">)</span><span class="token punctuation">,</span> ncol <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span>

<span class="token comment"># compute correlation matrix</span>
C <span class="token operator">&lt;-</span> cor<span class="token punctuation">(</span>mat<span class="token punctuation">)</span>

<span class="token comment"># plot correlations</span>
plotCorrMatrix<span class="token punctuation">(</span>C<span class="token punctuation">)</span>

<span class="token comment"># plot squared correlations</span>
plotCorrMatrix<span class="token punctuation">(</span>C<span class="token operator">^</span><span class="token number">2</span><span class="token punctuation">,</span> dendrogram <span class="token operator">=</span> <span class="token string">"none"</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotCorrStructure"><div class="page-main">
<a href="#plotCorrStructure" class="help-page-title"><h2>plotCorrStructure</h2></a>

<h3>Description</h3>

<p>Plot correlation structure of a gene based on random effects
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotCorrStructure(
  fit,
  varNames = names(coef(fit)),
  reorder = TRUE,
  pal = colorRampPalette(c("white", "red", "darkred")),
  hclust.method = "complete"
)
</code><code class="language-r">plotCorrStructure<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  varNames <span class="token operator">=</span> names<span class="token punctuation">(</span>coef<span class="token punctuation">(</span>fit<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  reorder <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  pal <span class="token operator">=</span> colorRampPalette<span class="token punctuation">(</span>c<span class="token punctuation">(</span><span class="token string">"white"</span><span class="token punctuation">,</span> <span class="token string">"red"</span><span class="token punctuation">,</span> <span class="token string">"darkred"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  hclust.method <span class="token operator">=</span> <span class="token string">"complete"</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>linear mixed model fit of a gene produced by lmer() or fitVarPartModel()</p>
</td>
</tr>
<tr>
<td><code id="varNames">varNames</code></td>
<td>
<p>variables in the metadata for which the correlation structure should be shown.  Variables must be random effects</p>
</td>
</tr>
<tr>
<td><code id="reorder">reorder</code></td>
<td>
<p>how to reorder the rows/columns of the correlation matrix.  reorder=FALSE gives no reorder.  reorder=TRUE reorders based on hclust.  reorder can also be an array of indices to reorder the samples manually</p>
</td>
</tr>
<tr>
<td><code id="pal">pal</code></td>
<td>
<p>color palette</p>
</td>
</tr>
<tr>
<td><code id="hclust.method">hclust.method</code></td>
<td>
<p>clustering methods for hclust</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>Image of correlation structure between each pair of experiments for a single gene
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
data(varPartData)

# specify formula
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# fit and return linear mixed models for each gene
fitList &lt;- fitVarPartModel(geneExpr[1:10, ], form, info)

# Focus on the first gene
fit &lt;- fitList[[1]]

# plot correlation sturcture based on Individual, reordering samples with hclust
plotCorrStructure(fit, "Individual")

# don't reorder
plotCorrStructure(fit, "Individual", reorder = FALSE)

# plot correlation sturcture based on Tissue, reordering samples with hclust
plotCorrStructure(fit, "Tissue")

# don't reorder
plotCorrStructure(fit, "Tissue", FALSE)

# plot correlation structure based on all random effects
# reorder manually by Tissue and Individual
idx &lt;- order(info$Tissue, info$Individual)
plotCorrStructure(fit, reorder = idx)

# plot correlation structure based on all random effects
# reorder manually by Individual, then Tissue
idx &lt;- order(info$Individual, info$Tissue)
plotCorrStructure(fit, reorder = idx)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># specify formula</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># fit and return linear mixed models for each gene</span>
fitList <span class="token operator">&lt;-</span> fitVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Focus on the first gene</span>
fit <span class="token operator">&lt;-</span> fitList<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span>

<span class="token comment"># plot correlation sturcture based on Individual, reordering samples with hclust</span>
plotCorrStructure<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token string">"Individual"</span><span class="token punctuation">)</span>

<span class="token comment"># don't reorder</span>
plotCorrStructure<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token string">"Individual"</span><span class="token punctuation">,</span> reorder <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span>

<span class="token comment"># plot correlation sturcture based on Tissue, reordering samples with hclust</span>
plotCorrStructure<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token string">"Tissue"</span><span class="token punctuation">)</span>

<span class="token comment"># don't reorder</span>
plotCorrStructure<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> <span class="token string">"Tissue"</span><span class="token punctuation">,</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span>

<span class="token comment"># plot correlation structure based on all random effects</span>
<span class="token comment"># reorder manually by Tissue and Individual</span>
idx <span class="token operator">&lt;-</span> order<span class="token punctuation">(</span>info<span class="token operator">$</span>Tissue<span class="token punctuation">,</span> info<span class="token operator">$</span>Individual<span class="token punctuation">)</span>
plotCorrStructure<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> reorder <span class="token operator">=</span> idx<span class="token punctuation">)</span>

<span class="token comment"># plot correlation structure based on all random effects</span>
<span class="token comment"># reorder manually by Individual, then Tissue</span>
idx <span class="token operator">&lt;-</span> order<span class="token punctuation">(</span>info<span class="token operator">$</span>Individual<span class="token punctuation">,</span> info<span class="token operator">$</span>Tissue<span class="token punctuation">)</span>
plotCorrStructure<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> reorder <span class="token operator">=</span> idx<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotPercentBars-method"><div class="page-main">
<a href="#plotPercentBars-method" class="help-page-title"><h2>Bar plot of gene fractions</h2></a>

<h3>Description</h3>

<p>Bar plot of fractions for a subset of genes
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotPercentBars(
  x,
  col = c(ggColorHue(ncol(x) - 1), "grey85"),
  genes = rownames(x),
  width = NULL,
  ...
)

## S4 method for signature 'matrix'
plotPercentBars(
  x,
  col = c(ggColorHue(ncol(x) - 1), "grey85"),
  genes = rownames(x),
  width = NULL,
  ...
)

## S4 method for signature 'data.frame'
plotPercentBars(
  x,
  col = c(ggColorHue(ncol(x) - 1), "grey85"),
  genes = rownames(x),
  width = NULL,
  ...
)

## S4 method for signature 'varPartResults'
plotPercentBars(
  x,
  col = c(ggColorHue(ncol(x) - 1), "grey85"),
  genes = rownames(x),
  width = NULL,
  ...
)
</code><code class="language-r">plotPercentBars<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>x<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  genes <span class="token operator">=</span> rownames<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">,</span>
  width <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'matrix'</span>
plotPercentBars<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>x<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  genes <span class="token operator">=</span> rownames<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">,</span>
  width <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'data.frame'</span>
plotPercentBars<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>x<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  genes <span class="token operator">=</span> rownames<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">,</span>
  width <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'varPartResults'</span>
plotPercentBars<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>x<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  genes <span class="token operator">=</span> rownames<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">,</span>
  width <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="x">x</code></td>
<td>
<p>object storing fractions</p>
</td>
</tr>
<tr>
<td><code id="col">col</code></td>
<td>
<p>color of bars for each variable</p>
</td>
</tr>
<tr>
<td><code id="genes">genes</code></td>
<td>
<p>name of genes to plot</p>
</td>
</tr>
<tr>
<td><code id="width">width</code></td>
<td>
<p>specify width of bars</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>Returns ggplot2 barplot
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Fit model
varPart &lt;- fitExtractVarPartModel(geneExpr, form, info)

# Bar plot for a subset of genes showing variance fractions
plotPercentBars(varPart[1:5, ])

# Move the legend to the top
plotPercentBars(varPart[1:5, ]) + theme(legend.position = "top")

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Fit model</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Bar plot for a subset of genes showing variance fractions</span>
plotPercentBars<span class="token punctuation">(</span>varPart<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">5</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">)</span>

<span class="token comment"># Move the legend to the top</span>
plotPercentBars<span class="token punctuation">(</span>varPart<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">5</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">)</span> <span class="token operator">+</span> theme<span class="token punctuation">(</span>legend.position <span class="token operator">=</span> <span class="token string">"top"</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotStratify"><div class="page-main">
<a href="#plotStratify" class="help-page-title"><h2>plotStratify</h2></a>

<h3>Description</h3>

<p>Plot gene expression stratified by another variable
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotStratify(
  formula,
  data,
  xlab,
  ylab,
  main,
  sortBy,
  colorBy,
  sort = TRUE,
  text = NULL,
  text.y = 1,
  text.size = 5,
  pts.cex = 1,
  ylim = NULL,
  legend = TRUE,
  x.labels = FALSE
)
</code><code class="language-r">plotStratify<span class="token punctuation">(</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  xlab<span class="token punctuation">,</span>
  ylab<span class="token punctuation">,</span>
  main<span class="token punctuation">,</span>
  sortBy<span class="token punctuation">,</span>
  colorBy<span class="token punctuation">,</span>
  sort <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  text <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  text.y <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  text.size <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">,</span>
  pts.cex <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  ylim <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  legend <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  x.labels <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specify variables shown in the x- and y-axes.  Y-axis should be continuous variable, x-axis should be discrete.</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p>data.frame storing continuous and discrete variables specified in formula</p>
</td>
</tr>
<tr>
<td><code id="xlab">xlab</code></td>
<td>
<p>label x-asis. Defaults to value of xval</p>
</td>
</tr>
<tr>
<td><code id="ylab">ylab</code></td>
<td>
<p>label y-asis. Defaults to value of yval</p>
</td>
</tr>
<tr>
<td><code id="main">main</code></td>
<td>
<p>main label</p>
</td>
</tr>
<tr>
<td><code id="sortBy">sortBy</code></td>
<td>
<p>name of column in geneExpr to sort samples by.  Defaults to xval</p>
</td>
</tr>
<tr>
<td><code id="colorBy">colorBy</code></td>
<td>
<p>name of column in geneExpr to color box plots.  Defaults to xval</p>
</td>
</tr>
<tr>
<td><code id="sort">sort</code></td>
<td>
<p>if TRUE, sort boxplots by median value, else use default ordering</p>
</td>
</tr>
<tr>
<td><code id="text">text</code></td>
<td>
<p>plot text on the top left of the plot</p>
</td>
</tr>
<tr>
<td><code id="text.y">text.y</code></td>
<td>
<p>indicate position of the text on the y-axis as a fraction of the y-axis range</p>
</td>
</tr>
<tr>
<td><code id="text.size">text.size</code></td>
<td>
<p>size of text</p>
</td>
</tr>
<tr>
<td><code id="pts.cex">pts.cex</code></td>
<td>
<p>size of points</p>
</td>
</tr>
<tr>
<td><code id="ylim">ylim</code></td>
<td>
<p>specify range of y-axis</p>
</td>
</tr>
<tr>
<td><code id="legend">legend</code></td>
<td>
<p>show legend</p>
</td>
</tr>
<tr>
<td><code id="x.labels">x.labels</code></td>
<td>
<p>show x axis labels</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>ggplot2 object
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# Note: This is a newer, more convient interface to plotStratifyBy()

# load library
# library(variancePartition)

# load simulated data:
data(varPartData)

# Create data.frame with expression and Tissue information for each sample
GE &lt;- data.frame(Expression = geneExpr[1, ], Tissue = info$Tissue)

# Plot expression stratified by Tissue
plotStratify(Expression ~ Tissue, GE)

# Omit legend and color boxes grey
plotStratify(Expression ~ Tissue, GE, colorBy = NULL)

# Specify colors
col &lt;- c(B = "green", A = "red", C = "yellow")
plotStratify(Expression ~ Tissue, GE, colorBy = col, sort = FALSE)

</code><code class="language-r"><span class="token comment"># Note: This is a newer, more convient interface to plotStratifyBy()</span>

<span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Create data.frame with expression and Tissue information for each sample</span>
GE <span class="token operator">&lt;-</span> data.frame<span class="token punctuation">(</span>Expression <span class="token operator">=</span> geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> Tissue <span class="token operator">=</span> info<span class="token operator">$</span>Tissue<span class="token punctuation">)</span>

<span class="token comment"># Plot expression stratified by Tissue</span>
plotStratify<span class="token punctuation">(</span>Expression <span class="token operator">~</span> Tissue<span class="token punctuation">,</span> GE<span class="token punctuation">)</span>

<span class="token comment"># Omit legend and color boxes grey</span>
plotStratify<span class="token punctuation">(</span>Expression <span class="token operator">~</span> Tissue<span class="token punctuation">,</span> GE<span class="token punctuation">,</span> colorBy <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">)</span>

<span class="token comment"># Specify colors</span>
col <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span>B <span class="token operator">=</span> <span class="token string">"green"</span><span class="token punctuation">,</span> A <span class="token operator">=</span> <span class="token string">"red"</span><span class="token punctuation">,</span> C <span class="token operator">=</span> <span class="token string">"yellow"</span><span class="token punctuation">)</span>
plotStratify<span class="token punctuation">(</span>Expression <span class="token operator">~</span> Tissue<span class="token punctuation">,</span> GE<span class="token punctuation">,</span> colorBy <span class="token operator">=</span> col<span class="token punctuation">,</span> sort <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotStratifyBy"><div class="page-main">
<a href="#plotStratifyBy" class="help-page-title"><h2>plotStratifyBy</h2></a>

<h3>Description</h3>

<p>Plot gene expression stratified by another variable
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotStratifyBy(
  geneExpr,
  xval,
  yval,
  xlab = xval,
  ylab = yval,
  main = NULL,
  sortBy = xval,
  colorBy = xval,
  sort = TRUE,
  text = NULL,
  text.y = 1,
  text.size = 5,
  pts.cex = 1,
  ylim = NULL,
  legend = TRUE,
  x.labels = FALSE
)
</code><code class="language-r">plotStratifyBy<span class="token punctuation">(</span>
  geneExpr<span class="token punctuation">,</span>
  xval<span class="token punctuation">,</span>
  yval<span class="token punctuation">,</span>
  xlab <span class="token operator">=</span> xval<span class="token punctuation">,</span>
  ylab <span class="token operator">=</span> yval<span class="token punctuation">,</span>
  main <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  sortBy <span class="token operator">=</span> xval<span class="token punctuation">,</span>
  colorBy <span class="token operator">=</span> xval<span class="token punctuation">,</span>
  sort <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  text <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  text.y <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  text.size <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">,</span>
  pts.cex <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  ylim <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  legend <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  x.labels <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="geneExpr">geneExpr</code></td>
<td>
<p>data.frame of gene expression values and another variable for each sample.  If there are multiple columns, the user can specify which one to use</p>
</td>
</tr>
<tr>
<td><code id="xval">xval</code></td>
<td>
<p>name of column in geneExpr to be used along x-axis to stratify gene expression</p>
</td>
</tr>
<tr>
<td><code id="yval">yval</code></td>
<td>
<p>name of column in geneExpr indicating gene expression</p>
</td>
</tr>
<tr>
<td><code id="xlab">xlab</code></td>
<td>
<p>label x-asis. Defaults to value of xval</p>
</td>
</tr>
<tr>
<td><code id="ylab">ylab</code></td>
<td>
<p>label y-asis. Defaults to value of yval</p>
</td>
</tr>
<tr>
<td><code id="main">main</code></td>
<td>
<p>main label</p>
</td>
</tr>
<tr>
<td><code id="sortBy">sortBy</code></td>
<td>
<p>name of column in geneExpr to sort samples by.  Defaults to xval</p>
</td>
</tr>
<tr>
<td><code id="colorBy">colorBy</code></td>
<td>
<p>name of column in geneExpr to color box plots.  Defaults to xval</p>
</td>
</tr>
<tr>
<td><code id="sort">sort</code></td>
<td>
<p>if TRUE, sort boxplots by median value, else use default ordering</p>
</td>
</tr>
<tr>
<td><code id="text">text</code></td>
<td>
<p>plot text on the top left of the plot</p>
</td>
</tr>
<tr>
<td><code id="text.y">text.y</code></td>
<td>
<p>indicate position of the text on the y-axis as a fraction of the y-axis range</p>
</td>
</tr>
<tr>
<td><code id="text.size">text.size</code></td>
<td>
<p>size of text</p>
</td>
</tr>
<tr>
<td><code id="pts.cex">pts.cex</code></td>
<td>
<p>size of points</p>
</td>
</tr>
<tr>
<td><code id="ylim">ylim</code></td>
<td>
<p>specify range of y-axis</p>
</td>
</tr>
<tr>
<td><code id="legend">legend</code></td>
<td>
<p>show legend</p>
</td>
</tr>
<tr>
<td><code id="x.labels">x.labels</code></td>
<td>
<p>show x axis labels</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>ggplot2 object
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

# load simulated data:
data(varPartData)

# Create data.frame with expression and Tissue information for each sample
GE &lt;- data.frame(Expression = geneExpr[1, ], Tissue = info$Tissue)

# Plot expression stratified by Tissue
plotStratifyBy(GE, "Tissue", "Expression")

# Omit legend and color boxes grey
plotStratifyBy(GE, "Tissue", "Expression", colorBy = NULL)

# Specify colors
col &lt;- c(B = "green", A = "red", C = "yellow")
plotStratifyBy(GE, "Tissue", "Expression", colorBy = col, sort = FALSE)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

<span class="token comment"># load simulated data:</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Create data.frame with expression and Tissue information for each sample</span>
GE <span class="token operator">&lt;-</span> data.frame<span class="token punctuation">(</span>Expression <span class="token operator">=</span> geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> Tissue <span class="token operator">=</span> info<span class="token operator">$</span>Tissue<span class="token punctuation">)</span>

<span class="token comment"># Plot expression stratified by Tissue</span>
plotStratifyBy<span class="token punctuation">(</span>GE<span class="token punctuation">,</span> <span class="token string">"Tissue"</span><span class="token punctuation">,</span> <span class="token string">"Expression"</span><span class="token punctuation">)</span>

<span class="token comment"># Omit legend and color boxes grey</span>
plotStratifyBy<span class="token punctuation">(</span>GE<span class="token punctuation">,</span> <span class="token string">"Tissue"</span><span class="token punctuation">,</span> <span class="token string">"Expression"</span><span class="token punctuation">,</span> colorBy <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">)</span>

<span class="token comment"># Specify colors</span>
col <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span>B <span class="token operator">=</span> <span class="token string">"green"</span><span class="token punctuation">,</span> A <span class="token operator">=</span> <span class="token string">"red"</span><span class="token punctuation">,</span> C <span class="token operator">=</span> <span class="token string">"yellow"</span><span class="token punctuation">)</span>
plotStratifyBy<span class="token punctuation">(</span>GE<span class="token punctuation">,</span> <span class="token string">"Tissue"</span><span class="token punctuation">,</span> <span class="token string">"Expression"</span><span class="token punctuation">,</span> colorBy <span class="token operator">=</span> col<span class="token punctuation">,</span> sort <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="plotVarianceEstimates"><div class="page-main">
<a href="#plotVarianceEstimates" class="help-page-title"><h2>Plot Variance Estimates</h2></a>

<h3>Description</h3>

<p>Plot Variance Estimates
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotVarianceEstimates(
  fit,
  fitEB,
  var_true = NULL,
  xmax = quantile(fit$sigma^2, 0.999)
)
</code><code class="language-r">plotVarianceEstimates<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  fitEB<span class="token punctuation">,</span>
  var_true <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  xmax <span class="token operator">=</span> quantile<span class="token punctuation">(</span>fit<span class="token operator">$</span>sigma<span class="token operator">^</span><span class="token number">2</span><span class="token punctuation">,</span> <span class="token number">0.999</span><span class="token punctuation">)</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit from <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="fitEB">fitEB</code></td>
<td>
<p>model fit from <code>eBayes()</code></p>
</td>
</tr>
<tr>
<td><code id="var_true">var_true</code></td>
<td>
<p>array of true variance values from simulation (optional)</p>
</td>
</tr>
<tr>
<td><code id="xmax">xmax</code></td>
<td>
<p>maximum value on the x-axis</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="plotVarPart-method"><div class="page-main">
<a href="#plotVarPart-method" class="help-page-title"><h2>Violin plot of variance fractions</h2></a>

<h3>Description</h3>

<p>Violin plot of variance fraction for each gene and each variable
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">plotVarPart(
  obj,
  col = c(ggColorHue(ncol(obj) - 1), "grey85"),
  label.angle = 20,
  main = "",
  ylab = "",
  convertToPercent = TRUE,
  ...
)

## S4 method for signature 'matrix'
plotVarPart(
  obj,
  col = c(ggColorHue(ncol(obj) - 1), "grey85"),
  label.angle = 20,
  main = "",
  ylab = "",
  convertToPercent = TRUE,
  ...
)

## S4 method for signature 'data.frame'
plotVarPart(
  obj,
  col = c(ggColorHue(ncol(obj) - 1), "grey85"),
  label.angle = 20,
  main = "",
  ylab = "",
  convertToPercent = TRUE,
  ...
)

## S4 method for signature 'varPartResults'
plotVarPart(
  obj,
  col = c(ggColorHue(ncol(obj) - 1), "grey85"),
  label.angle = 20,
  main = "",
  ylab = "",
  convertToPercent = TRUE,
  ...
)
</code><code class="language-r">plotVarPart<span class="token punctuation">(</span>
  obj<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>obj<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  label.angle <span class="token operator">=</span> <span class="token number">20</span><span class="token punctuation">,</span>
  main <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  ylab <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  convertToPercent <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'matrix'</span>
plotVarPart<span class="token punctuation">(</span>
  obj<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>obj<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  label.angle <span class="token operator">=</span> <span class="token number">20</span><span class="token punctuation">,</span>
  main <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  ylab <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  convertToPercent <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'data.frame'</span>
plotVarPart<span class="token punctuation">(</span>
  obj<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>obj<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  label.angle <span class="token operator">=</span> <span class="token number">20</span><span class="token punctuation">,</span>
  main <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  ylab <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  convertToPercent <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'varPartResults'</span>
plotVarPart<span class="token punctuation">(</span>
  obj<span class="token punctuation">,</span>
  col <span class="token operator">=</span> c<span class="token punctuation">(</span>ggColorHue<span class="token punctuation">(</span>ncol<span class="token punctuation">(</span>obj<span class="token punctuation">)</span> <span class="token operator">-</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token string">"grey85"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  label.angle <span class="token operator">=</span> <span class="token number">20</span><span class="token punctuation">,</span>
  main <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  ylab <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span>
  convertToPercent <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="obj">obj</code></td>
<td>
<p><code>varParFrac</code> object returned by <code>fitExtractVarPart</code> or <code>extractVarPart</code></p>
</td>
</tr>
<tr>
<td><code id="col">col</code></td>
<td>
<p>vector of colors</p>
</td>
</tr>
<tr>
<td><code id="label.angle">label.angle</code></td>
<td>
<p>angle of labels on x-axis</p>
</td>
</tr>
<tr>
<td><code id="main">main</code></td>
<td>
<p>title of plot</p>
</td>
</tr>
<tr>
<td><code id="ylab">ylab</code></td>
<td>
<p>text on y-axis</p>
</td>
</tr>
<tr>
<td><code id="convertToPercent">convertToPercent</code></td>
<td>
<p>multiply fractions by 100 to convert to percent values</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>additional arguments</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>Makes violin plots of variance components model.  This function uses the graphics interface from ggplot2.  Warnings produced by this function usually ggplot2 warning that the window is too small.
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

varPart &lt;- fitExtractVarPartModel(geneExpr, form, info)

# violin plot of contribution of each variable to total variance
plotVarPart(sortCols(varPart))

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># violin plot of contribution of each variable to total variance</span>
plotVarPart<span class="token punctuation">(</span>sortCols<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="rdf"><div class="page-main">
<a href="#rdf" class="help-page-title"><h2>Residual degrees of freedom</h2></a>

<h3>Description</h3>

<p>Residual degrees of freedom
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">rdf(fit)
</code><code class="language-r">rdf<span class="token punctuation">(</span>fit<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation"><tr>
<td><code id="fit">fit</code></td>
<td>
<p>model fit from <code>lm()</code>, <code>glm()</code>, <code>lmer()</code></p>
</td>
</tr></table>
<h3>See Also</h3>

<p><code>rdf.merMod</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">library(lme4)

fit &lt;- lm(Reaction ~ Days, sleepstudy)
rdf(fit)

</code><code class="language-r">library<span class="token punctuation">(</span>lme4<span class="token punctuation">)</span>

fit <span class="token operator">&lt;-</span> lm<span class="token punctuation">(</span>Reaction <span class="token operator">~</span> Days<span class="token punctuation">,</span> sleepstudy<span class="token punctuation">)</span>
rdf<span class="token punctuation">(</span>fit<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="rdf_from_matrices"><div class="page-main">
<a href="#rdf_from_matrices" class="help-page-title"><h2>Fast approximate residual degrees of freedom</h2></a>

<h3>Description</h3>

<p>Defining <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mo>=</mo><msup><mi>A</mi><mi>T</mi></msup><mi>A</mi><mo>+</mo><msup><mi>B</mi><mi>T</mi></msup><mi>B</mi></mrow><annotation encoding="application/x-tex">H = A^TA + B^TB</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.9247em;vertical-align:-0.0833em;"></span><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span><span class="mord mathnormal">A</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.8413em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span></span></span></span></code> where <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi></mrow><annotation encoding="application/x-tex">A</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal">A</span></span></span></span></code> and <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi></mrow><annotation encoding="application/x-tex">B</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span></span></span></span></code> are low rank, compute
<code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>−</mo><mn>2</mn><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mi>H</mi><mo stretchy="false">)</mo><mo>+</mo><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mi>H</mi><mi>H</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">n - 2tr(H) + tr(HH)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6667em;vertical-align:-0.0833em;"></span><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">2</span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">HH</span><span class="mclose">)</span></span></span></span></code> in <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mo stretchy="false">(</mo><mi>n</mi><msup><mi>p</mi><mn>2</mn></msup><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">O(np^2)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.0641em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">O</span><span class="mopen">(</span><span class="mord mathnormal">n</span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></code> instead of <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mo stretchy="false">(</mo><msup><mi>n</mi><mn>2</mn></msup><msup><mi>p</mi><mn>2</mn></msup><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">O(n^2p^2)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.0641em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">O</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">n</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></code>.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">rdf_from_matrices(A, B)
</code><code class="language-r">rdf_from_matrices<span class="token punctuation">(</span>A<span class="token punctuation">,</span> B<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="A">A</code></td>
<td>
<p>a <code>matrix</code> or <code>sparseMatrix</code></p>
</td>
</tr>
<tr>
<td><code id="B">B</code></td>
<td>
<p>a <code>matrix</code> or <code>sparseMatrix</code></p>
</td>
</tr>
</table>
<h3>See Also</h3>

<p>rdf.merMod
</p>

<hr>
</div></div>
<div class="container manual-page" id="rdf.merMod"><div class="page-main">
<a href="#rdf.merMod" class="help-page-title"><h2>Approximate residual degrees of freedom</h2></a>

<h3>Description</h3>

<p>For a linear model with <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi></mrow><annotation encoding="application/x-tex">n</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">n</span></span></span></span></code> samples and <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi></mrow><annotation encoding="application/x-tex">p</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.1944em;"></span><span class="mord mathnormal">p</span></span></span></span></code> covariates, <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>R</mi><mi>S</mi><mi>S</mi><mi mathvariant="normal">/</mi><msup><mi>σ</mi><mn>2</mn></msup><mo>∼</mo><msubsup><mi>χ</mi><mi>ν</mi><mn>2</mn></msubsup></mrow><annotation encoding="application/x-tex">RSS/\sigma^2 \sim \chi^2_{\nu}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.0641em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.05764em;">RSS</span><span class="mord">/</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">σ</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">∼</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1.0611em;vertical-align:-0.247em;"></span><span class="mord"><span class="mord mathnormal">χ</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-2.453em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.06366em;">ν</span></span></span></span><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.247em;"><span></span></span></span></span></span></span></span></span></span></code> where <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ν</mi><mo>=</mo><mi>n</mi><mo>−</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">\nu = n-p</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal" style="margin-right:0.06366em;">ν</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6667em;vertical-align:-0.0833em;"></span><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.1944em;"></span><span class="mord mathnormal">p</span></span></span></span></code> is the residual degrees of freedom.  In the case of a linear mixed model, the distribution is no longer exactly a chi-square distribution, but can be approximated with a chi-square distribution.
</p>
<p>Given the hat matrix, <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi></mrow><annotation encoding="application/x-tex">H</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span></span></span></span></code>, that maps between observed and fitted responses, the approximate residual degrees of freedom is <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ν</mi><mo>=</mo><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mo stretchy="false">(</mo><mi>I</mi><mo>−</mo><mi>H</mi><msup><mo stretchy="false">)</mo><mi>T</mi></msup><mo stretchy="false">(</mo><mi>I</mi><mo>−</mo><mi>H</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">\nu = tr((I-H)^T(I-H))</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal" style="margin-right:0.06366em;">ν</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">((</span><span class="mord mathnormal" style="margin-right:0.07847em;">I</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1.0913em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.07847em;">I</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mclose">))</span></span></span></span></code>.  For a linear model, this simplifies to the well known form <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ν</mi><mo>=</mo><mi>n</mi><mo>−</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">\nu = n - p</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal" style="margin-right:0.06366em;">ν</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6667em;vertical-align:-0.0833em;"></span><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.1944em;"></span><span class="mord mathnormal">p</span></span></span></span></code>. In the more general case, such as a linear mixed model, the original form simplifies only to <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>−</mo><mn>2</mn><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mi>H</mi><mo stretchy="false">)</mo><mo>+</mo><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mi>H</mi><mi>H</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">n - 2tr(H) + tr(HH)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6667em;vertical-align:-0.0833em;"></span><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">2</span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">HH</span><span class="mclose">)</span></span></span></span></code> and is an approximation rather than being exact.  The third term here is quadratic time in the number of samples, <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi></mrow><annotation encoding="application/x-tex">n</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">n</span></span></span></span></code>, and can be computationally expensive to evaluate for larger datasets.  Here we develop a linear time algorithm that takes advantage of the fact that <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi></mrow><annotation encoding="application/x-tex">H</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span></span></span></span></code> is low rank.
</p>
<p><code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi></mrow><annotation encoding="application/x-tex">H</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span></span></span></span></code> is computed as <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>A</mi><mi>T</mi></msup><mi>A</mi><mo>+</mo><msup><mi>B</mi><mi>T</mi></msup><mi>B</mi></mrow><annotation encoding="application/x-tex">A^TA + B^TB</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.9247em;vertical-align:-0.0833em;"></span><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span><span class="mord mathnormal">A</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.8413em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span></span></span></span></code> for <code>A=CL</code> and <code>B=CR</code> defined in the code.  Since <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi></mrow><annotation encoding="application/x-tex">A</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal">A</span></span></span></span></code> and <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi></mrow><annotation encoding="application/x-tex">B</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span></span></span></span></code> are low rank, there is no need to compute <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi></mrow><annotation encoding="application/x-tex">H</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span></span></span></span></code> directly.  Instead, the terms <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mi>H</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">tr(H)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mclose">)</span></span></span></span></code> and <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mi>r</mi><mo stretchy="false">(</mo><mi>H</mi><mi>H</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">tr(HH)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">t</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">HH</span><span class="mclose">)</span></span></span></span></code> can be computed using the eigen decompositions of <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><msup><mi>A</mi><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">AA^T</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8413em;"></span><span class="mord mathnormal">A</span><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span></span></span></span></code> and <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi><msup><mi>B</mi><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">BB^T</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8413em;"></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8413em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">T</span></span></span></span></span></span></span></span></span></span></span></code> which is linear time in the number of samples.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">rdf.merMod(model, method = c("linear", "quadratic"))
</code><code class="language-r">rdf.merMod<span class="token punctuation">(</span>model<span class="token punctuation">,</span> method <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"linear"</span><span class="token punctuation">,</span> <span class="token string">"quadratic"</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="model">model</code></td>
<td>
<p>An object of class <code>merMod</code></p>
</td>
</tr>
<tr>
<td><code id="method">method</code></td>
<td>
<p>Use algorithm that is "linear" (default) or quadratic time in the number of samples</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Compute the approximate residual degrees of freedom from a linear mixed model.
</p>


<h3>Value</h3>

<p>residual degrees of freedom
</p>


<h3>See Also</h3>

<p>rdf_from_matrices
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">library(lme4)

# Fit linear mixed model
fit &lt;- lmer(Reaction ~ Days + (Days | Subject), sleepstudy)

# Evaluate the approximate residual degrees of freedom
rdf.merMod(fit)

</code><code class="language-r">library<span class="token punctuation">(</span>lme4<span class="token punctuation">)</span>

<span class="token comment"># Fit linear mixed model</span>
fit <span class="token operator">&lt;-</span> lmer<span class="token punctuation">(</span>Reaction <span class="token operator">~</span> Days <span class="token operator">+</span> <span class="token punctuation">(</span>Days <span class="token operator">|</span> Subject<span class="token punctuation">)</span><span class="token punctuation">,</span> sleepstudy<span class="token punctuation">)</span>

<span class="token comment"># Evaluate the approximate residual degrees of freedom</span>
rdf.merMod<span class="token punctuation">(</span>fit<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="reOnly"><div class="page-main">
<a href="#reOnly" class="help-page-title"><h2>Adapted from lme4:::reOnly</h2></a>

<h3>Description</h3>

<p>Adapted from lme4:::reOnly
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">reOnly(f, response = FALSE)
</code><code class="language-r">reOnly<span class="token punctuation">(</span>f<span class="token punctuation">,</span> response <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="f">f</code></td>
<td>
<p>formula</p>
</td>
</tr>
<tr>
<td><code id="response">response</code></td>
<td>
<p>(FALSE) is there a response in the formula</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="residuals-MArrayLM-method"><div class="page-main">
<a href="#residuals-MArrayLM-method" class="help-page-title"><h2>residuals for MArrayLM</h2></a>

<h3>Description</h3>

<p>residuals for MArrayLM
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM'
residuals(object, y, ..., type = c("response", "pearson"))
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM'</span>
residuals<span class="token punctuation">(</span>object<span class="token punctuation">,</span> y<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">,</span> type <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"response"</span><span class="token punctuation">,</span> <span class="token string">"pearson"</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>MArrayLM object from dream</p>
</td>
</tr>
<tr>
<td><code id="y">y</code></td>
<td>
<p><code>EList</code> object used in <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments, currently ignored</p>
</td>
</tr>
<tr>
<td><code id="type">type</code></td>
<td>
<p>compute either response or pearson residuals</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>results of residuals
</p>

<hr>
</div></div>
<div class="container manual-page" id="residuals-MArrayLM2-method"><div class="page-main">
<a href="#residuals-MArrayLM2-method" class="help-page-title"><h2>residuals for MArrayLM2</h2></a>

<h3>Description</h3>

<p>residuals for MArrayLM2
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM2'
residuals(object, y, type = c("response", "pearson"), ...)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM2'</span>
residuals<span class="token punctuation">(</span>object<span class="token punctuation">,</span> y<span class="token punctuation">,</span> type <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"response"</span><span class="token punctuation">,</span> <span class="token string">"pearson"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>MArrayLM2 object from dream</p>
</td>
</tr>
<tr>
<td><code id="y">y</code></td>
<td>
<p><code>EList</code> object used in <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="type">type</code></td>
<td>
<p>compute either response or pearson residuals</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments, currently ignored</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>results of residuals
</p>

<hr>
</div></div>
<div class="container manual-page" id="residuals-VarParFitList-method"><div class="page-main">
<a href="#residuals-VarParFitList-method" class="help-page-title"><h2>Residuals from model fit</h2></a>

<h3>Description</h3>

<p>Extract residuals for each gene from model fit with fitVarPartModel()
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'VarParFitList'
residuals(object, ...)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'VarParFitList'</span>
residuals<span class="token punctuation">(</span>object<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>object produced by fitVarPartModel()</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments.</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>If model is fit with missing data, residuals returns NA for entries that were
missing in the original data
</p>


<h3>Value</h3>

<p>Residuals extracted from model fits stored in object
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Fit model
modelFit &lt;- fitVarPartModel(geneExpr, form, info)

# Extract residuals of model fit
res &lt;- residuals(modelFit)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Fit model</span>
modelFit <span class="token operator">&lt;-</span> fitVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># Extract residuals of model fit</span>
res <span class="token operator">&lt;-</span> residuals<span class="token punctuation">(</span>modelFit<span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="residuals.MArrayLM2"><div class="page-main">
<a href="#residuals.MArrayLM2" class="help-page-title"><h2>Residuals for result of dream</h2></a>

<h3>Description</h3>

<p>Residuals for result of dream
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">residuals.MArrayLM2(object, y, ..., type = c("response", "pearson"))
</code><code class="language-r">residuals.MArrayLM2<span class="token punctuation">(</span>object<span class="token punctuation">,</span> y<span class="token punctuation">,</span> <span class="token ellipsis">...</span><span class="token punctuation">,</span> type <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"response"</span><span class="token punctuation">,</span> <span class="token string">"pearson"</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p>See <code>?stats::residuals</code></p>
</td>
</tr>
<tr>
<td><code id="y">y</code></td>
<td>
<p><code>EList</code> object used in <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>See <code>?stats::residuals</code></p>
</td>
</tr>
<tr>
<td><code id="type">type</code></td>
<td>
<p>compute either response or pearson residuals</p>
</td>
</tr>
</table>
<hr>
</div></div>
<div class="container manual-page" id="shrinkageMetric"><div class="page-main">
<a href="#shrinkageMetric" class="help-page-title"><h2>Shrinkage metric for eBayes</h2></a>

<h3>Description</h3>

<p>Evaluates the coefficient from the linear regression of <code>s2.post ~ sigmaSq</code>. When there is no shrinkage, this value is 1.  Values less than 1 indicate the amount of shrinkage.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">shrinkageMetric(sigmaSq, s2.post)
</code><code class="language-r">shrinkageMetric<span class="token punctuation">(</span>sigmaSq<span class="token punctuation">,</span> s2.post<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="sigmaSq">sigmaSq</code></td>
<td>
<p>maximum likelihood residual variance for every gene</p>
</td>
</tr>
<tr>
<td><code id="s2.post">s2.post</code></td>
<td>
<p>empirical Bayes posterior estimate of residual variance for every gene</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Shrinkage metric for eBayes quantifying the amount of shrinkage that is applied to shrink the maximum likelihood residual variance to the empirical Bayes posterior estimate
</p>

<hr>
</div></div>
<div class="container manual-page" id="sortCols-method"><div class="page-main">
<a href="#sortCols-method" class="help-page-title"><h2>Sort variance partition statistics</h2></a>

<h3>Description</h3>

<p>Sort columns returned by <code>extractVarPart()</code> or <code>fitExtractVarPartModel()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">sortCols(
  x,
  FUN = median,
  decreasing = TRUE,
  last = c("Residuals", "Measurement.error"),
  ...
)

## S4 method for signature 'matrix'
sortCols(
  x,
  FUN = median,
  decreasing = TRUE,
  last = c("Residuals", "Measurement.error"),
  ...
)

## S4 method for signature 'data.frame'
sortCols(
  x,
  FUN = median,
  decreasing = TRUE,
  last = c("Residuals", "Measurement.error"),
  ...
)

## S4 method for signature 'varPartResults'
sortCols(
  x,
  FUN = median,
  decreasing = TRUE,
  last = c("Residuals", "Measurement.error"),
  ...
)
</code><code class="language-r">sortCols<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  FUN <span class="token operator">=</span> median<span class="token punctuation">,</span>
  decreasing <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  last <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"Residuals"</span><span class="token punctuation">,</span> <span class="token string">"Measurement.error"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'matrix'</span>
sortCols<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  FUN <span class="token operator">=</span> median<span class="token punctuation">,</span>
  decreasing <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  last <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"Residuals"</span><span class="token punctuation">,</span> <span class="token string">"Measurement.error"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'data.frame'</span>
sortCols<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  FUN <span class="token operator">=</span> median<span class="token punctuation">,</span>
  decreasing <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  last <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"Residuals"</span><span class="token punctuation">,</span> <span class="token string">"Measurement.error"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'varPartResults'</span>
sortCols<span class="token punctuation">(</span>
  x<span class="token punctuation">,</span>
  FUN <span class="token operator">=</span> median<span class="token punctuation">,</span>
  decreasing <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  last <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"Residuals"</span><span class="token punctuation">,</span> <span class="token string">"Measurement.error"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="x">x</code></td>
<td>
<p>object returned by <code>extractVarPart()</code> or <code>fitExtractVarPartModel()</code></p>
</td>
</tr>
<tr>
<td><code id="FUN">FUN</code></td>
<td>
<p>function giving summary statistic to sort by.  Defaults to median</p>
</td>
</tr>
<tr>
<td><code id="decreasing">decreasing</code></td>
<td>
<p>logical.  Should the sorting be increasing or decreasing?</p>
</td>
</tr>
<tr>
<td><code id="last">last</code></td>
<td>
<p>columns to be placed on the right, regardless of values in these columns</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments to sort</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>data.frame with columns sorted by mean value, with Residuals in last column
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Step 1: fit linear mixed model on gene expression
# If categorical variables are specified, a linear mixed model is used
# If all variables are modeled as continuous, a linear model is used
# each entry in results is a regression model fit on a single gene
# Step 2: extract variance fractions from each model fit
# for each gene, returns fraction of variation attributable to each variable
# Interpretation: the variance explained by each variable
# after correction for all other variables
varPart &lt;- fitExtractVarPartModel(geneExpr, form, info)

# violin plot of contribution of each variable to total variance
# sort columns by median value
plotVarPart(sortCols(varPart))

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Step 1: fit linear mixed model on gene expression</span>
<span class="token comment"># If categorical variables are specified, a linear mixed model is used</span>
<span class="token comment"># If all variables are modeled as continuous, a linear model is used</span>
<span class="token comment"># each entry in results is a regression model fit on a single gene</span>
<span class="token comment"># Step 2: extract variance fractions from each model fit</span>
<span class="token comment"># for each gene, returns fraction of variation attributable to each variable</span>
<span class="token comment"># Interpretation: the variance explained by each variable</span>
<span class="token comment"># after correction for all other variables</span>
varPart <span class="token operator">&lt;-</span> fitExtractVarPartModel<span class="token punctuation">(</span>geneExpr<span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>

<span class="token comment"># violin plot of contribution of each variable to total variance</span>
<span class="token comment"># sort columns by median value</span>
plotVarPart<span class="token punctuation">(</span>sortCols<span class="token punctuation">(</span>varPart<span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="toptable-method"><div class="page-main">
<a href="#toptable-method" class="help-page-title"><h2>Table of Top Genes from Linear Model Fit</h2></a>

<h3>Description</h3>

<p>topTable generic
</p>
<p>topTable generic MArrayLM
</p>
<p>topTable generic MArrayLM2
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">topTable(
  fit,
  coef = NULL,
  number = 10,
  genelist = fit$genes,
  adjust.method = "BH",
  sort.by = "B",
  resort.by = NULL,
  p.value = 1,
  lfc = 0,
  confint = FALSE
)

## S4 method for signature 'MArrayLM'
topTable(
  fit,
  coef = NULL,
  number = 10,
  genelist = fit$genes,
  adjust.method = "BH",
  sort.by = "p",
  resort.by = NULL,
  p.value = 1,
  lfc = 0,
  confint = FALSE
)

## S4 method for signature 'MArrayLM2'
topTable(
  fit,
  coef = NULL,
  number = 10,
  genelist = fit$genes,
  adjust.method = "BH",
  sort.by = "p",
  resort.by = NULL,
  p.value = 1,
  lfc = 0,
  confint = FALSE
)
</code><code class="language-r">topTable<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  coef <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span>
  genelist <span class="token operator">=</span> fit<span class="token operator">$</span>genes<span class="token punctuation">,</span>
  adjust.method <span class="token operator">=</span> <span class="token string">"BH"</span><span class="token punctuation">,</span>
  sort.by <span class="token operator">=</span> <span class="token string">"B"</span><span class="token punctuation">,</span>
  resort.by <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  p.value <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  lfc <span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">,</span>
  confint <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM'</span>
topTable<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  coef <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span>
  genelist <span class="token operator">=</span> fit<span class="token operator">$</span>genes<span class="token punctuation">,</span>
  adjust.method <span class="token operator">=</span> <span class="token string">"BH"</span><span class="token punctuation">,</span>
  sort.by <span class="token operator">=</span> <span class="token string">"p"</span><span class="token punctuation">,</span>
  resort.by <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  p.value <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  lfc <span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">,</span>
  confint <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM2'</span>
topTable<span class="token punctuation">(</span>
  fit<span class="token punctuation">,</span>
  coef <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span>
  genelist <span class="token operator">=</span> fit<span class="token operator">$</span>genes<span class="token punctuation">,</span>
  adjust.method <span class="token operator">=</span> <span class="token string">"BH"</span><span class="token punctuation">,</span>
  sort.by <span class="token operator">=</span> <span class="token string">"p"</span><span class="token punctuation">,</span>
  resort.by <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  p.value <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>
  lfc <span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">,</span>
  confint <span class="token operator">=</span> <span class="token boolean">FALSE</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="fit">fit</code></td>
<td>
<p>fit</p>
</td>
</tr>
<tr>
<td><code id="coef">coef</code></td>
<td>
<p>coef</p>
</td>
</tr>
<tr>
<td><code id="number">number</code></td>
<td>
<p>number</p>
</td>
</tr>
<tr>
<td><code id="genelist">genelist</code></td>
<td>
<p>genelist</p>
</td>
</tr>
<tr>
<td><code id="adjust.method">adjust.method</code></td>
<td>
<p>adjust.method</p>
</td>
</tr>
<tr>
<td><code id="sort.by">sort.by</code></td>
<td>
<p>sort.by</p>
</td>
</tr>
<tr>
<td><code id="resort.by">resort.by</code></td>
<td>
<p>resort.by</p>
</td>
</tr>
<tr>
<td><code id="p.value">p.value</code></td>
<td>
<p>p.value</p>
</td>
</tr>
<tr>
<td><code id="lfc">lfc</code></td>
<td>
<p>lfc</p>
</td>
</tr>
<tr>
<td><code id="confint">confint</code></td>
<td>
<p>confint</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>results of toptable
</p>
<p>results of toptable
</p>
<p>results of toptable
</p>

<hr>
</div></div>
<div class="container manual-page" id="VarParCIList-class"><div class="page-main">
<a href="#VarParCIList-class" class="help-page-title"><h2>Class VarParCIList</h2></a>

<h3>Description</h3>

<p>Class <code>VarParCIList</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="VarParFitList-class"><div class="page-main">
<a href="#VarParFitList-class" class="help-page-title"><h2>Class VarParFitList</h2></a>

<h3>Description</h3>

<p>Class <code>VarParFitList</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="varParFrac-class"><div class="page-main">
<a href="#varParFrac-class" class="help-page-title"><h2>Class varParFrac</h2></a>

<h3>Description</h3>

<p>Class <code>varParFrac</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="varPartConfInf"><div class="page-main">
<a href="#varPartConfInf" class="help-page-title"><h2>Linear mixed model confidence intervals</h2></a>

<h3>Description</h3>

<p>Fit linear mixed model to estimate contribution of multiple sources of variation while simultaneously correcting for all other variables. Then perform parametric bootstrap sampling to get a 95% confidence intervals for each variable for each gene.
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">varPartConfInf(
  exprObj,
  formula,
  data,
  REML = FALSE,
  useWeights = TRUE,
  control = vpcontrol,
  nsim = 1000,
  ...
)
</code><code class="language-r">varPartConfInf<span class="token punctuation">(</span>
  exprObj<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  REML <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  useWeights <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  control <span class="token operator">=</span> vpcontrol<span class="token punctuation">,</span>
  nsim <span class="token operator">=</span> <span class="token number">1000</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="exprObj">exprObj</code></td>
<td>
<p>matrix of expression data (g genes x n samples), or <code>ExpressionSet</code>, or <code>EList</code> returned by <code>voom()</code> from the <code>limma</code> package</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code></p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p><code>data.frame</code> with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="REML">REML</code></td>
<td>
<p>use restricted maximum likelihood to fit linear mixed model. default is FALSE.  Strongly discourage against changing this option, but here for compatibility.</p>
</td>
</tr>
<tr>
<td><code id="useWeights">useWeights</code></td>
<td>
<p>if TRUE, analysis uses heteroskedastic error estimates from <code>voom()</code>.  Value is ignored unless exprObj is an <code>EList</code> from <code>voom()</code> or <code>weightsMatrix</code> is specified</p>
</td>
</tr>
<tr>
<td><code id="control">control</code></td>
<td>
<p>control settings for <code>lmer()</code></p>
</td>
</tr>
<tr>
<td><code id="nsim">nsim</code></td>
<td>
<p>number of bootstrap datasets</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>Additional arguments for <code>lmer()</code> or l<code>m()</code></p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>A linear mixed model is fit for each gene, and <code>bootMer()</code> is used to generate parametric bootstrap confidence intervals.  <code>use.u=TRUE</code> is used so that the <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover accent="true"><mi>u</mi><mo>^</mo></mover></mrow><annotation encoding="application/x-tex">\hat{u}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em;"></span><span class="mord accent"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.6944em;"><span style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="mord mathnormal">u</span></span><span style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="accent-body" style="left:-0.2222em;"><span class="mord">^</span></span></span></span></span></span></span></span></span></span></code> values from the random effects are used as estimated and are not re-sampled.  This gives confidence intervals as if additional data were generated from these same current samples.  Conversely, <code>use.u=FALSE</code> assumes that  this dataset is a sample from a larger population.   Thus it simulates <code class="reqn"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover accent="true"><mi>u</mi><mo>^</mo></mover></mrow><annotation encoding="application/x-tex">\hat{u}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em;"></span><span class="mord accent"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.6944em;"><span style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="mord mathnormal">u</span></span><span style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="accent-body" style="left:-0.2222em;"><span class="mord">^</span></span></span></span></span></span></span></span></span></span></code> based on the estimated variance parameter.  This approach gives confidence intervals as if additional data were collected from the larger population from which this dataset is sampled.  Overall, <code>use.u=TRUE</code> gives smaller confidence intervals that are appropriate in this case.
</p>


<h3>Value</h3>

<p><code>list()</code> of where each entry is the result for a gene.  Each entry is a matrix of the 95% confidence interval of the variance fraction for each variable
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">
# load library
# library(variancePartition)

library(BiocParallel)

# load simulated data:
# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
data(varPartData)

# Specify variables to consider
# Age is continuous so we model it as a fixed effect
# Individual and Tissue are both categorical, so we model them as random effects
form &lt;- ~ Age + (1 | Individual) + (1 | Tissue)

# Compute bootstrap confidence intervals for each variable for each gene
resCI &lt;- varPartConfInf(geneExpr[1:5, ], form, info, nsim = 100)

</code><code class="language-r"><span class="token comment"># load library</span>
<span class="token comment"># library(variancePartition)</span>

library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

<span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

<span class="token comment"># Specify variables to consider</span>
<span class="token comment"># Age is continuous so we model it as a fixed effect</span>
<span class="token comment"># Individual and Tissue are both categorical, so we model them as random effects</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Age <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Tissue<span class="token punctuation">)</span>

<span class="token comment"># Compute bootstrap confidence intervals for each variable for each gene</span>
resCI <span class="token operator">&lt;-</span> varPartConfInf<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">5</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">,</span> nsim <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="varPartData"><div class="page-main">
<a href="#varPartData" class="help-page-title"><h2>Simulation dataset for examples</h2></a>

<h3>Description</h3>

<p>A simulated dataset of gene expression and metadata
</p>
<p>A simulated dataset of gene counts
</p>
<p>A simulated dataset of gene counts
</p>
<p>A simulated dataset of gene counts
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">data(varPartData)

data(varPartData)

data(varPartData)

data(varPartData)
</code><code class="language-r">data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span></code></pre>


<h3>Format</h3>

<p>A dataset of 100 samples and 200 genes
</p>
<p>A dataset of 100 samples and 200 genes
</p>
<p>A dataset of 100 samples and 200 genes
</p>
<p>A dataset of 100 samples and 200 genes
</p>


<h3>Details</h3>


<ul>
<li>
<p> geneCounts gene expression in the form of RNA-seq counts
</p>
</li>
<li>
<p> geneExpr gene expression on a continuous scale
</p>
</li>
<li>
<p> info metadata about the study design
</p>
</li>
</ul>
<ul>
<li>
<p> geneCounts gene expression in the form of RNA-seq counts
</p>
</li>
<li>
<p> geneExpr gene expression on a continuous scale
</p>
</li>
<li>
<p> info metadata about the study design
</p>
</li>
</ul>
<ul>
<li>
<p> geneCounts gene expression in the form of RNA-seq counts
</p>
</li>
<li>
<p> geneExpr gene expression on a continuous scale
</p>
</li>
<li>
<p> info metadata about the study design
</p>
</li>
</ul>
<ul>
<li>
<p> geneCounts gene expression in the form of RNA-seq counts
</p>
</li>
<li>
<p> geneExpr gene expression on a continuous scale
</p>
</li>
<li>
<p> info metadata about the study design
</p>
</li>
</ul>
<hr>
</div></div>
<div class="container manual-page" id="varPartDEdata"><div class="page-main">
<a href="#varPartDEdata" class="help-page-title"><h2>A simulated dataset of gene counts</h2></a>

<h3>Description</h3>


<ul>
<li>
<p> geneCounts gene expression in the form of RNA-seq counts
</p>
</li>
<li>
<p> geneExpr gene expression on a continuous scale
</p>
</li>
<li>
<p> info metadata about the study design
</p>
</li>
</ul>
<ul>
<li>
<p> geneCounts gene expression in the form of RNA-seq counts
</p>
</li>
<li>
<p> geneExpr gene expression on a continuous scale
</p>
</li>
<li>
<p> info metadata about the study design
</p>
</li>
</ul>
<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">data(varPartData)

data(varPartData)
</code><code class="language-r">data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span></code></pre>


<h3>Format</h3>

<p>A dataset of 24 samples and 19,364 genes
</p>
<p>A dataset of 24 samples and 19,364 genes
</p>

<hr>
</div></div>
<div class="container manual-page" id="varPartResults-class"><div class="page-main">
<a href="#varPartResults-class" class="help-page-title"><h2>Class varPartResults</h2></a>

<h3>Description</h3>

<p>Class <code>varPartResults</code>
</p>

<hr>
</div></div>
<div class="container manual-page" id="vcov-MArrayLM-method"><div class="page-main">
<a href="#vcov-MArrayLM-method" class="help-page-title"><h2>Co-variance matrix for <code>dream()</code> fit</h2></a>

<h3>Description</h3>

<p>Define generic <code>vcov()</code> for result of <code>lmFit()</code> and <code>dream()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM'
vcov(object, vobj, coef)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM'</span>
vcov<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> coef<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p><code>MArrayLM</code> object return by <code>lmFit()</code> or <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> object returned by <code>voom()</code></p>
</td>
</tr>
<tr>
<td><code id="coef">coef</code></td>
<td>
<p>name of coefficient to be extracted</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>variance-covariance matrix
</p>

<hr>
</div></div>
<div class="container manual-page" id="vcov-MArrayLM2-method"><div class="page-main">
<a href="#vcov-MArrayLM2-method" class="help-page-title"><h2>Co-variance matrix for <code>dream()</code> fit</h2></a>

<h3>Description</h3>

<p>Define generic <code>vcov()</code> for result of <code>lmFit()</code> and <code>dream()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">## S4 method for signature 'MArrayLM2'
vcov(object, vobj, coef)
</code><code class="language-r"><span class="token comment">## S4 method for signature 'MArrayLM2'</span>
vcov<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> coef<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p><code>MArrayLM</code> object return by <code>lmFit()</code> or <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> object returned by <code>voom()</code></p>
</td>
</tr>
<tr>
<td><code id="coef">coef</code></td>
<td>
<p>name of coefficient to be extracted</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>variance-covariance matrix
</p>

<hr>
</div></div>
<div class="container manual-page" id="vcovSqrt-method"><div class="page-main">
<a href="#vcovSqrt-method" class="help-page-title"><h2>Sqrt of co-variance matrix for <code>dream()</code> fit</h2></a>

<h3>Description</h3>

<p>Define generic <code>vcovSqrt()</code> for result of <code>lmFit()</code> and <code>dream()</code>
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">vcovSqrt(object, vobj, coef, approx = TRUE)

## S4 method for signature 'MArrayLM'
vcovSqrt(object, vobj, coef, approx = TRUE)

## S4 method for signature 'MArrayLM2'
vcovSqrt(object, vobj, coef, approx = TRUE)
</code><code class="language-r">vcovSqrt<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> coef<span class="token punctuation">,</span> approx <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM'</span>
vcovSqrt<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> coef<span class="token punctuation">,</span> approx <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>

<span class="token comment">## S4 method for signature 'MArrayLM2'</span>
vcovSqrt<span class="token punctuation">(</span>object<span class="token punctuation">,</span> vobj<span class="token punctuation">,</span> coef<span class="token punctuation">,</span> approx <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="object">object</code></td>
<td>
<p><code>MArrayLM</code> object return by <code>lmFit()</code> or <code>dream()</code></p>
</td>
</tr>
<tr>
<td><code id="vobj">vobj</code></td>
<td>
<p><code>EList</code> object returned by <code>voom()</code></p>
</td>
</tr>
<tr>
<td><code id="coef">coef</code></td>
<td>
<p>name of coefficient to be extracted</p>
</td>
</tr>
<tr>
<td><code id="approx">approx</code></td>
<td>
<p>use fast approximation</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>Computes factor of covariance matrix so that <code>vcov(object)</code> is the same as <code>crossprod(vcovSqrt(object))</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># load simulated data:
# geneExpr: matrix of *normalized* gene expression values
# info: information/metadata about each sample
data(varPartData)

form &lt;- ~Batch

fit &lt;- dream(geneExpr[1:2, ], form, info)
fit &lt;- eBayes(fit)

# Compute covariance directly
Sigma &lt;- vcov(fit, geneExpr[1:2, ])

# Compute factor of covariance
S &lt;- crossprod(vcovSqrt(fit, geneExpr[1:2, ]))

</code><code class="language-r"><span class="token comment"># load simulated data:</span>
<span class="token comment"># geneExpr: matrix of *normalized* gene expression values</span>
<span class="token comment"># info: information/metadata about each sample</span>
data<span class="token punctuation">(</span>varPartData<span class="token punctuation">)</span>

form <span class="token operator">&lt;-</span> <span class="token operator">~</span>Batch

fit <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">2</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> info<span class="token punctuation">)</span>
fit <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>fit<span class="token punctuation">)</span>

<span class="token comment"># Compute covariance directly</span>
Sigma <span class="token operator">&lt;-</span> vcov<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">2</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">)</span>

<span class="token comment"># Compute factor of covariance</span>
S <span class="token operator">&lt;-</span> crossprod<span class="token punctuation">(</span>vcovSqrt<span class="token punctuation">(</span>fit<span class="token punctuation">,</span> geneExpr<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">2</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
<div class="container manual-page" id="voomWithDreamWeights"><div class="page-main">
<a href="#voomWithDreamWeights" class="help-page-title"><h2>Transform RNA-Seq Data Ready for Linear Mixed Modelling with <code>dream()</code>
</h2></a>

<h3>Description</h3>

<p>Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with <code>dream()</code>.   This method is the same as <code>limma::voom()</code>, except that it allows random effects in the formula
</p>


<h3>Usage</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden">voomWithDreamWeights(
  counts,
  formula,
  data,
  lib.size = NULL,
  normalize.method = "none",
  span = 0.5,
  weights = NULL,
  prior.count = 0.5,
  prior.count.for.weights = prior.count,
  plot = FALSE,
  save.plot = TRUE,
  rescaleWeightsAfter = FALSE,
  scaledByLib = FALSE,
  priorWeightsAsCounts = FALSE,
  BPPARAM = SerialParam(),
  ...
)
</code><code class="language-r">voomWithDreamWeights<span class="token punctuation">(</span>
  counts<span class="token punctuation">,</span>
  formula<span class="token punctuation">,</span>
  data<span class="token punctuation">,</span>
  lib.size <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  normalize.method <span class="token operator">=</span> <span class="token string">"none"</span><span class="token punctuation">,</span>
  span <span class="token operator">=</span> <span class="token number">0.5</span><span class="token punctuation">,</span>
  weights <span class="token operator">=</span> <span class="token keyword">NULL</span><span class="token punctuation">,</span>
  prior.count <span class="token operator">=</span> <span class="token number">0.5</span><span class="token punctuation">,</span>
  prior.count.<span class="token keyword">for</span>.weights <span class="token operator">=</span> prior.count<span class="token punctuation">,</span>
  plot <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  save.plot <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>
  rescaleWeightsAfter <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  scaledByLib <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  priorWeightsAsCounts <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span>
  BPPARAM <span class="token operator">=</span> SerialParam<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token ellipsis">...</span>
<span class="token punctuation">)</span></code></pre>


<h3 class="r-arguments-title">Arguments</h3>

<table role="presentation">
<tr>
<td><code id="counts">counts</code></td>
<td>
<p>a numeric <code>matrix</code> containing raw counts, or an <code>ExpressionSet</code> containing raw counts, or a <code>DGEList</code> object. Counts must be non-negative and NAs are not permitted.</p>
</td>
</tr>
<tr>
<td><code id="formula">formula</code></td>
<td>
<p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work, and <code>lmFit()</code> followed by contrasts.fit() are run.</p>
</td>
</tr>
<tr>
<td><code id="data">data</code></td>
<td>
<p><code>data.frame</code> with columns corresponding to formula</p>
</td>
</tr>
<tr>
<td><code id="lib.size">lib.size</code></td>
<td>
<p>numeric vector containing total library sizes for each sample.  Defaults to the normalized (effective) library sizes in <code>counts</code> if <code>counts</code> is a <code>DGEList</code> or to the columnwise count totals if <code>counts</code> is a matrix.</p>
</td>
</tr>
<tr>
<td><code id="normalize.method">normalize.method</code></td>
<td>
<p>the microarray-style normalization method to be applied to the logCPM values (if any).  Choices are as for the <code>method</code> argument of <code>normalizeBetweenArrays</code> when the data is single-channel.  Any normalization factors found in <code>counts</code> will still be used even if <code>normalize.method="none"</code>.</p>
</td>
</tr>
<tr>
<td><code id="span">span</code></td>
<td>
<p>width of the lowess smoothing window as a proportion. Setting <code>span="auto"</code> uses <code>fANCOVA::loess.as()</code> to estimate the tuning parameter from the data</p>
</td>
</tr>
<tr>
<td><code id="weights">weights</code></td>
<td>
<p>Can be a numeric matrix of individual weights of same dimensions as the <code>counts</code>, or a numeric vector of sample weights with length equal to <code>ncol(counts)</code></p>
</td>
</tr>
<tr>
<td><code id="prior.count">prior.count</code></td>
<td>
<p>average count to be added to each observation to avoid taking log of zero. The count applied to each sample is normalized by library size so given equal log CPM for a gene with zero counts across multiple samples</p>
</td>
</tr>
<tr>
<td><code id="prior.count.for.weights">prior.count.for.weights</code></td>
<td>
<p>count added to regularize weights</p>
</td>
</tr>
<tr>
<td><code id="plot">plot</code></td>
<td>
<p>logical, should a plot of the mean-variance trend be displayed?</p>
</td>
</tr>
<tr>
<td><code id="save.plot">save.plot</code></td>
<td>
<p>logical, should the coordinates and line of the plot be saved in the output?</p>
</td>
</tr>
<tr>
<td><code id="rescaleWeightsAfter">rescaleWeightsAfter</code></td>
<td>
<p>default = FALSE, should the output weights be scaled by the input weights</p>
</td>
</tr>
<tr>
<td><code id="scaledByLib">scaledByLib</code></td>
<td>
<p>if <code>TRUE</code>, scale pseudocount by <code>lib.size</code>.  Else to standard constant pseudocount addition</p>
</td>
</tr>
<tr>
<td><code id="priorWeightsAsCounts">priorWeightsAsCounts</code></td>
<td>
<p>if <code>weights</code> is <code>NULL</code>, set weights to be equal to counts, following delta method for log2 CPM</p>
</td>
</tr>
<tr>
<td><code id="BPPARAM">BPPARAM</code></td>
<td>
<p>parameters for parallel evaluation</p>
</td>
</tr>
<tr>
<td><code id="...">...</code></td>
<td>
<p>other arguments are passed to <code>lmer</code>.</p>
</td>
</tr>
</table>
<h3>Details</h3>

<p>Adapted from <code>voom()</code> in <code>limma</code> v3.40.2
</p>


<h3>Value</h3>

<p>An <code>EList</code> object just like the result of <code>limma::voom()</code>
</p>


<h3>See Also</h3>

<p><code>limma::voom()</code>
</p>


<h3>Examples</h3>

<pre class="language-r"><code class="language-r-input" hidden="hidden"># library(variancePartition)
library(edgeR)
library(BiocParallel)

data(varPartDEdata)

# normalize RNA-seq counts
dge &lt;- DGEList(counts = countMatrix)
dge &lt;- calcNormFactors(dge)

# specify formula with random effect for Individual
form &lt;- ~ Disease + (1 | Individual)

# compute observation weights
vobj &lt;- voomWithDreamWeights(dge[1:20, ], form, metadata)

# fit dream model
res &lt;- dream(vobj, form, metadata)
res &lt;- eBayes(res)

# extract results
topTable(res, coef = "Disease1", number = 3)

</code><code class="language-r"><span class="token comment"># library(variancePartition)</span>
library<span class="token punctuation">(</span>edgeR<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>

data<span class="token punctuation">(</span>varPartDEdata<span class="token punctuation">)</span>

<span class="token comment"># normalize RNA-seq counts</span>
dge <span class="token operator">&lt;-</span> DGEList<span class="token punctuation">(</span>counts <span class="token operator">=</span> countMatrix<span class="token punctuation">)</span>
dge <span class="token operator">&lt;-</span> calcNormFactors<span class="token punctuation">(</span>dge<span class="token punctuation">)</span>

<span class="token comment"># specify formula with random effect for Individual</span>
form <span class="token operator">&lt;-</span> <span class="token operator">~</span> Disease <span class="token operator">+</span> <span class="token punctuation">(</span><span class="token number">1</span> <span class="token operator">|</span> Individual<span class="token punctuation">)</span>

<span class="token comment"># compute observation weights</span>
vobj <span class="token operator">&lt;-</span> voomWithDreamWeights<span class="token punctuation">(</span>dge<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">20</span><span class="token punctuation">,</span> <span class="token punctuation">]</span><span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>

<span class="token comment"># fit dream model</span>
res <span class="token operator">&lt;-</span> dream<span class="token punctuation">(</span>vobj<span class="token punctuation">,</span> form<span class="token punctuation">,</span> metadata<span class="token punctuation">)</span>
res <span class="token operator">&lt;-</span> eBayes<span class="token punctuation">(</span>res<span class="token punctuation">)</span>

<span class="token comment"># extract results</span>
topTable<span class="token punctuation">(</span>res<span class="token punctuation">,</span> coef <span class="token operator">=</span> <span class="token string">"Disease1"</span><span class="token punctuation">,</span> number <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span></code></pre>

<hr>
</div></div>
</div>
    <footer><p>Rendered with postdoc 1.4.1</p></footer>
  <script data-cfasync="false" src="/cdn-cgi/scripts/5c5dd728/cloudflare-static/email-decode.min.js"></script><script defer src="https://static.cloudflareinsights.com/beacon.min.js/vcd15cbe7772f49c399c6a5babf22c1241717689176015" integrity="sha512-ZpsOmlRQV6y907TI0dKBHq9Md29nnaEIPlkf84rnaERnq6zvWvPUqr2ft8M1aS28oN72PdrCzSjY4U6VaAw1EQ==" data-cf-beacon='{"version":"2024.11.0","token":"6000532f154d4405a1c967fdafb41d60","r":1,"server_timing":{"name":{"cfCacheStatus":true,"cfEdge":true,"cfExtPri":true,"cfL4":true,"cfOrigin":true,"cfSpeedBrain":true},"location_startswith":null}}' crossorigin="anonymous"></script>
</body>
</html>
